<!DOCTYPE html><html class="theme-next mist" lang="zh-Hans"><head><meta name="generator" content="Hexo 3.8.0"><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width,initial-scale=1,maximum-scale=1"><meta name="theme-color" content="#222"><script src="/lib/pace/pace.min.js?v=1.0.2"></script><link href="/lib/pace/pace-theme-minimal.min.css?v=1.0.2" rel="stylesheet"><script></script><meta http-equiv="Cache-Control" content="no-transform"><meta http-equiv="Cache-Control" content="no-siteapp"><link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css"><link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css"><link href="/css/main.css?v=5.1.4" rel="stylesheet" type="text/css"><link rel="apple-touch-icon" sizes="180x180" href="/images/dark_logo_180x180.jpg?v=5.1.4"><link rel="icon" type="image/png" sizes="32x32" href="/images/dark_logo_32x32.jpg?v=5.1.4"><link rel="icon" type="image/png" sizes="16x16" href="/images/dark_logo_16x16.jpg?v=5.1.4"><link rel="mask-icon" href="/images/dark_logo.svg?v=5.1.4" color="#222"><meta name="keywords" content="绘图,"><link rel="alternate" href="/atom.xml" title="showteeth's blog" type="application/atom+xml"><meta name="description" content="这篇文章主要学习了使用ComplexHeatmap来绘制想要的热图的过程，首先以绘制差异表达基因的热图为例学习了常规热图的绘制方法，后续又详细学习了使用oncoPrint绘制突变谱的方法。"><meta name="keywords" content="绘图"><meta property="og:type" content="article"><meta property="og:title" content="R绘图系列-ComplexHeatmap使用记录"><meta property="og:url" content="http://showteeth.tech/posts/31202.html"><meta property="og:site_name" content="showteeth&#39;s blog"><meta property="og:description" content="这篇文章主要学习了使用ComplexHeatmap来绘制想要的热图的过程，首先以绘制差异表达基因的热图为例学习了常规热图的绘制方法，后续又详细学习了使用oncoPrint绘制突变谱的方法。"><meta property="og:locale" content="zh-Hans"><meta property="og:image" content="https://showteeth.oss-cn-beijing.aliyuncs.com/blog_img/complexheatmap_order_row_and_split.png"><meta property="og:image" content="https://showteeth.oss-cn-beijing.aliyuncs.com/blog_img/complexheatmap_anno_col_row_names.png"><meta property="og:image" content="https://showteeth.oss-cn-beijing.aliyuncs.com/blog_img/complexheatmap_anno_block.png"><meta property="og:image" content="https://showteeth.oss-cn-beijing.aliyuncs.com/blog_img/complexheatmap_anno_block_split.png"><meta property="og:image" content="https://showteeth.oss-cn-beijing.aliyuncs.com/blog_img/complexheatmap_oncoprint_grid.png"><meta property="og:image" content="https://showteeth.oss-cn-beijing.aliyuncs.com/blog_img/complexheatmap_oncoprint_grid_2.png"><meta property="og:image" content="https://showteeth.oss-cn-beijing.aliyuncs.com/blog_img/complexheatmap_oncoprint_pro_1.png"><meta property="og:image" content="https://showteeth.oss-cn-beijing.aliyuncs.com/blog_img/complexheatmap_oncoprint_tri.png"><meta property="og:image" content="https://showteeth.oss-cn-beijing.aliyuncs.com/blog_img/complexheatmap_oncoprint_test_alter_1.png"><meta property="og:image" content="https://showteeth.oss-cn-beijing.aliyuncs.com/blog_img/complexheatmap_oncoprint_bg_2.png"><meta property="og:image" content="https://showteeth.oss-cn-beijing.aliyuncs.com/blog_img/complexheatmap_oncoprint_bg_1.png"><meta property="og:image" content="https://showteeth.oss-cn-beijing.aliyuncs.com/blog_img/complexheatmap_oncoprint_intron_exon_1.png"><meta property="og:image" content="https://showteeth.oss-cn-beijing.aliyuncs.com/blog_img/complexheatmap_oncoprint_intron_exon_2.png"><meta property="og:image" content="https://showteeth.oss-cn-beijing.aliyuncs.com/blog_img/complexheatmap_oncoprint_intron_exon_3.png"><meta property="og:image" content="https://showteeth.oss-cn-beijing.aliyuncs.com/blog_img/complexheatmap_oncoprint_complex_1.png"><meta property="og:updated_time" content="2020-05-18T11:13:19.690Z"><meta name="twitter:card" content="summary"><meta name="twitter:title" content="R绘图系列-ComplexHeatmap使用记录"><meta name="twitter:description" content="这篇文章主要学习了使用ComplexHeatmap来绘制想要的热图的过程，首先以绘制差异表达基因的热图为例学习了常规热图的绘制方法，后续又详细学习了使用oncoPrint绘制突变谱的方法。"><meta name="twitter:image" content="https://showteeth.oss-cn-beijing.aliyuncs.com/blog_img/complexheatmap_order_row_and_split.png"><script type="text/javascript" id="hexo.configurations">var NexT=window.NexT||{},CONFIG={root:"/",scheme:"Mist",version:"5.1.4",sidebar:{position:"left",display:"post",offset:12,b2t:!1,scrollpercent:!0,onmobile:!1},fancybox:!0,tabs:!0,motion:{enable:!1,async:!1,transition:{post_block:"fadeIn",post_header:"slideDownIn",post_body:"slideDownIn",coll_header:"slideLeftIn",sidebar:"slideUpIn"}},duoshuo:{userId:"0",author:"博主"},algolia:{applicationID:"",apiKey:"",indexName:"",hits:{per_page:10},labels:{input_placeholder:"Search for Posts",hits_empty:"We didn't find any results for the search: ${query}",hits_stats:"${hits} results found in ${time} ms"}}}</script><link rel="canonical" href="http://showteeth.tech/posts/31202.html"><link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css"><script>!function(e,t,o,c,i,a,n){e.DaoVoiceObject=i,e[i]=e[i]||function(){(e[i].q=e[i].q||[]).push(arguments)},e[i].l=1*new Date,a=t.createElement(o),n=t.getElementsByTagName(o)[0],a.async=1,a.src=c,a.charset="utf-8",n.parentNode.insertBefore(a,n)}(window,document,"script",("https:"==document.location.protocol?"https:":"http:")+"//widget.daovoice.io/widget/9d02ffb0.js","daovoice"),daovoice("init",{app_id:"9d02ffb0"}),daovoice("update")</script><title>R绘图系列-ComplexHeatmap使用记录 | showteeth's blog</title></head><body itemscope itemtype="http://schema.org/WebPage" lang="zh-Hans"><div class="container sidebar-position-left page-post-detail"><div class="headband"></div><a href="https://github.com/showteeth" class="github-corner" aria-label="View source on GitHub"><svg width="80" height="80" viewbox="0 0 250 250" style="fill:#151513;color:#fff;position:absolute;top:0;border:0;right:0" aria-hidden="true"><path d="M0,0 L115,115 L130,115 L142,142 L250,250 L250,0 Z"/><path d="M128.3,109.0 C113.8,99.7 119.0,89.6 119.0,89.6 C122.0,82.7 120.5,78.6 120.5,78.6 C119.2,72.0 123.4,76.3 123.4,76.3 C127.3,80.9 125.5,87.3 125.5,87.3 C122.9,97.6 130.6,101.9 134.4,103.2" fill="currentColor" style="transform-origin:130px 106px" class="octo-arm"/><path d="M115.0,115.0 C114.9,115.1 118.7,116.5 119.8,115.4 L133.7,101.6 C136.9,99.2 139.9,98.4 142.2,98.6 C133.8,88.0 127.5,74.4 143.8,58.0 C148.5,53.4 154.0,51.2 159.7,51.0 C160.3,49.4 163.2,43.6 171.4,40.1 C171.4,40.1 176.1,42.5 178.8,56.2 C183.1,58.6 187.2,61.8 190.9,65.4 C194.5,69.0 197.7,73.2 200.1,77.6 C213.8,80.2 216.3,84.9 216.3,84.9 C212.7,93.1 206.9,96.0 205.4,96.6 C205.1,102.4 203.0,107.8 198.3,112.5 C181.9,128.9 168.3,122.5 157.7,114.1 C157.9,116.9 156.7,120.9 152.7,124.9 L141.0,136.5 C139.8,137.7 141.6,141.9 141.8,141.8 Z" fill="currentColor" class="octo-body"/></svg></a><style>.github-corner:hover .octo-arm{animation:octocat-wave 560ms ease-in-out}@keyframes octocat-wave{0%,100%{transform:rotate(0)}20%,60%{transform:rotate(-25deg)}40%,80%{transform:rotate(10deg)}}@media (max-width:500px){.github-corner:hover .octo-arm{animation:none}.github-corner .octo-arm{animation:octocat-wave 560ms ease-in-out}}</style><header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader"><div class="header-inner"><div class="site-brand-wrapper"><div class="site-meta"><div class="custom-logo-site-title"><a href="/" class="brand" rel="start"><span class="logo-line-before"><i></i></span> <span class="site-title">showteeth's blog</span> <span class="logo-line-after"><i></i></span></a></div><p class="site-subtitle"></p></div><div class="site-nav-toggle"><button><span class="btn-bar"></span> <span class="btn-bar"></span> <span class="btn-bar"></span></button></div></div><nav class="site-nav"><ul id="menu" class="menu"><li class="menu-item menu-item-home"><a href="/" rel="section"><i class="menu-item-icon fa fa-fw fa-home"></i><br>首页</a></li><li class="menu-item menu-item-archives"><a href="/archives/" rel="section"><i class="menu-item-icon fa fa-fw fa-archive"></i><br>归档</a></li><li class="menu-item menu-item-categories"><a href="/categories/" rel="section"><i class="menu-item-icon fa fa-fw fa-th"></i><br>分类</a></li><li class="menu-item menu-item-tags"><a href="/tags/" rel="section"><i class="menu-item-icon fa fa-fw fa-tags"></i><br>标签</a></li><li class="menu-item menu-item-photos"><a href="/photos" rel="section"><i class="menu-item-icon fa fa-fw fa-camera"></i><br>相册</a></li><li class="menu-item menu-item-about"><a href="/about/" rel="section"><i class="menu-item-icon fa fa-fw fa-user"></i><br>关于</a></li><li class="menu-item menu-item-search"><a href="javascript:;" class="popup-trigger"><i class="menu-item-icon fa fa-search fa-fw"></i><br>搜索</a></li></ul><div class="site-search"><div class="popup search-popup local-search-popup"><div class="local-search-header clearfix"><span class="search-icon"><i class="fa fa-search"></i> </span><span class="popup-btn-close"><i class="fa fa-times-circle"></i></span><div class="local-search-input-wrapper"><input autocomplete="off" placeholder="搜索..." spellcheck="false" type="text" id="local-search-input"></div></div><div id="local-search-result"></div></div></div></nav></div></header><main id="main" class="main"><div class="main-inner"><div class="content-wrap"><div id="content" class="content"><div id="posts" class="posts-expand"><article class="post post-type-normal" itemscope itemtype="http://schema.org/Article"><div class="post-block"><link itemprop="mainEntityOfPage" href="http://showteeth.tech/posts/31202.html"><span hidden itemprop="author" itemscope itemtype="http://schema.org/Person"><meta itemprop="name" content="showteeth"><meta itemprop="description" content><meta itemprop="image" content="/images/timg.jfif"></span><span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization"><meta itemprop="name" content="showteeth's blog"></span><header class="post-header"><h1 class="post-title" itemprop="name headline">R绘图系列-ComplexHeatmap使用记录</h1><div class="post-meta"><span class="post-time"><span class="post-meta-item-icon"><i class="fa fa-calendar-o"></i> </span><span class="post-meta-item-text">发表于</span> <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-12-19T21:13:09+08:00">2019-12-19 </time><span class="post-meta-divider">|</span> <span class="post-meta-item-icon"><i class="fa fa-calendar-check-o"></i> </span><span class="post-meta-item-text">更新于&#58;</span> <time title="更新于" itemprop="dateModified" datetime="2020-05-18T19:13:19+08:00">2020-05-18 </time></span><span class="post-category"><span class="post-meta-divider">|</span> <span class="post-meta-item-icon"><i class="fa fa-folder-o"></i> </span><span class="post-meta-item-text">分类于</span> <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/R/" itemprop="url" rel="index"><span itemprop="name">R</span> </a></span>， <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/R/绘图/" itemprop="url" rel="index"><span itemprop="name">绘图</span> </a></span></span><span class="post-comments-count"><span class="post-meta-divider">|</span> <span class="post-meta-item-icon"><i class="fa fa-comment-o"></i> </span><a href="/posts/31202.html#comments" itemprop="discussionUrl"><span class="post-comments-count valine-comment-count" data-xid="/posts/31202.html" itemprop="commentCount"></span> </a></span><span class="post-meta-divider">|</span> <span class="page-pv"><i class="fa fa-file-o"></i> 本文总阅读量 <span class="busuanzi-value" id="busuanzi_value_page_pv"></span>次</span><div class="post-wordcount"><span class="post-meta-item-icon"><i class="fa fa-file-word-o"></i> </span><span class="post-meta-item-text">字数统计&#58;</span> <span title="字数统计">5.6k 字 </span><span class="post-meta-divider">|</span> <span class="post-meta-item-icon"><i class="fa fa-clock-o"></i> </span><span class="post-meta-item-text">阅读时长 &asymp;</span> <span title="阅读时长">26 分钟</span></div></div></header><div class="post-body" itemprop="articleBody"><div class="note info"><p>这篇文章主要学习了使用<code>ComplexHeatmap</code>来绘制想要的热图的过程，首先以<strong>绘制差异表达基因的热图</strong>为例学习了常规热图的绘制方法，后续又<strong>详细学习了使用<code>oncoPrint</code>绘制突变谱的方法</strong>。</p></div><a id="more"></a><h2 id="输入数据类型"><a href="#输入数据类型" class="headerlink" title="输入数据类型"></a>输入数据类型</h2><p><strong>最好是矩阵</strong>，如果是<code>dataframe</code>可能会出现<code>warning</code>：<br></p><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">Warning message:</span><br><span class="line">The input is a data frame, convert it to the matrix.</span><br></pre></td></tr></table></figure><p></p><p>如果想<strong>颠倒横纵坐标画图</strong>，可以直接使用<code>t(matrix)</code>进行转置。</p><hr><h2 id="设置颜色"><a href="#设置颜色" class="headerlink" title="设置颜色"></a>设置颜色</h2><p>设置颜色有多种方式，可以使用<code>ComplexHeatmap</code>推荐的<code>colorRamp2</code>函数，其优点在于<strong>可以设置不同的breaks对应的颜色值</strong>：<br></p><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">library</span>(circlize)</span><br><span class="line">red_green=colorRamp2(breaks=c(min(diff_genes_matrix), </span><br><span class="line">                       mean(diff_genes_matrix), </span><br><span class="line">                       max(diff_genes_matrix)),</span><br><span class="line">                       colors=c(<span class="string">"red"</span>, <span class="string">"black"</span>, <span class="string">"green"</span>))</span><br></pre></td></tr></table></figure><p></p><hr><h2 id="设置legend"><a href="#设置legend" class="headerlink" title="设置legend"></a>设置legend</h2><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">heatmap_legend_param = list(color_bar = <span class="string">"continuous"</span>, </span><br><span class="line">                            at = quantile(diff_genes_matrix, c(<span class="number">0</span>,<span class="number">1</span>)), </span><br><span class="line">                            labels = c(<span class="string">"Min"</span>, <span class="string">"Max"</span>),</span><br><span class="line">                            legend_height = unit(<span class="number">4</span>, <span class="string">"cm"</span>),</span><br><span class="line">                            title_position = <span class="string">"leftcenter-rot"</span>)</span><br></pre></td></tr></table></figure><ul><li><code>color_bar</code>: 连续的</li><li><code>at</code>和后面的<code>labels</code>连用，在什么地方标注什么</li><li><code>legend_height</code>设置legend高度</li><li><code>title_position</code>：设置<code>legend</code>名称的位置，这个需要和<code>Heatmap</code>的参数<code>name</code>连用来控制<code>legend</code>名称，在垂直的<code>legend</code>下可用的包括<code>&#39;topleft&#39;</code>, <code>&#39;topcenter&#39;</code>, <code>&#39;leftcenter-rot&#39;</code> and <code>&#39;lefttop-rot&#39;</code></li><li><code>legend_direction = &quot;horizontal&quot;</code>：设置legend横置</li></ul><p>参考链接：</p><ul><li><a href="https://jokergoo.github.io/blog/html/more_legends.html" target="_blank" rel="noopener">ComplexHeatmap Supports More Types of Legends</a></li><li><a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/legends.html" target="_blank" rel="noopener">Heatmap and annotation legends</a></li></ul><hr><h2 id="设置row和col名称的顺序"><a href="#设置row和col名称的顺序" class="headerlink" title="设置row和col名称的顺序"></a>设置row和col名称的顺序</h2><p>在设置<code>row</code>和<code>col</code>名称的顺序时，<strong>行和列的聚类自动会去除</strong>：<br></p><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># row1、row10、row11、row2...</span></span><br><span class="line">row_order = sort(rownames(diff_genes_matrix))</span><br><span class="line"></span><br><span class="line"><span class="comment"># 直接指定顺序：</span></span><br><span class="line">my_order=c(<span class="string">"row1"</span>,<span class="string">"row2"</span>,<span class="string">"row10"</span>,<span class="string">"row11"</span>)</span><br><span class="line">row_order = my_order</span><br></pre></td></tr></table></figure><p></p><p>参考链接：</p><ul><li><a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/a-single-heatmap.html#row-and_column_orders" target="_blank" rel="noopener">row-and_column_orders</a></li></ul><p>需要注意的是：<mark>如果在<strong>设置了顺序的同时也设置了<code>row_split</code></strong>，那么<strong>手动设置的<code>row_order</code>就会失效</strong>；在设置了<code>cluster_rows = FALSE</code>的情况下，默认的顺序就是<code>matrix</code>行名的顺序</mark>，参考<a href="https://github.com/jokergoo/ComplexHeatmap/issues/299" target="_blank" rel="noopener">How to avoid reordering of matrix components with Split?</a>，所以正确的做法是：</p><ul><li>先修改<code>matrix</code>的顺序，可以参考<a href="http://showteeth.tech/posts/30224.html">R系列之向量、矩阵、数组、数据框和列表</a></li><li>关闭<code>cluster_rows</code>参数，直接绘图即可<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 原始的矩阵顺序</span></span><br><span class="line">rownames(diff_genes_matrix)</span><br><span class="line">        [<span class="number">1</span>] <span class="string">"BE_W_1"</span>    <span class="string">"BE_W_3"</span>    <span class="string">"BE_W_6"</span>    <span class="string">"BE_C_2"</span>    <span class="string">"BE_C_4"</span>    <span class="string">"BE_C_5"</span>    <span class="string">"Mutation1"</span> <span class="string">"Mutation2"</span> <span class="string">"Mutation3"</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 先修改matrix顺序</span></span><br><span class="line">diff_genes_matrix_sort=diff_genes_matrix[c(<span class="string">"BE_W_1"</span>,<span class="string">"BE_W_3"</span>,<span class="string">"BE_W_6"</span>,<span class="string">"Mutation1"</span>,<span class="string">"Mutation2"</span>,<span class="string">"Mutation3"</span>,<span class="string">"BE_C_2"</span>,<span class="string">"BE_C_4"</span>,<span class="string">"BE_C_5"</span>),]</span><br><span class="line">rownames(diff_genes_matrix_sort)</span><br><span class="line">        [<span class="number">1</span>] <span class="string">"BE_W_1"</span>    <span class="string">"BE_W_3"</span>    <span class="string">"BE_W_6"</span>    <span class="string">"Mutation1"</span> <span class="string">"Mutation2"</span> <span class="string">"Mutation3"</span> <span class="string">"BE_C_2"</span>    <span class="string">"BE_C_4"</span>    <span class="string">"BE_C_5"</span></span><br><span class="line"></span><br><span class="line">Heatmap(diff_genes_matrix_sort, show_column_names = <span class="literal">FALSE</span>, </span><br><span class="line">        show_row_dend = <span class="literal">FALSE</span>, show_column_dend = <span class="literal">F</span>, </span><br><span class="line">        col=red_green,</span><br><span class="line">        name=<span class="string">"Expression level"</span>,</span><br><span class="line">        </span><br><span class="line">        row_names_side = <span class="string">"left"</span>,</span><br><span class="line">        cluster_rows = <span class="literal">FALSE</span>,</span><br><span class="line">        heatmap_width = unit(<span class="number">24</span>, <span class="string">"cm"</span>), heatmap_height = unit(<span class="number">8</span>, <span class="string">"cm"</span>),</span><br><span class="line">        row_split = split,</span><br><span class="line">        row_title = <span class="literal">NULL</span>,</span><br><span class="line">        row_gap = unit(<span class="number">0.5</span>, <span class="string">"mm"</span>),</span><br><span class="line">        heatmap_legend_param = list(color_bar = <span class="string">"continuous"</span>, </span><br><span class="line">                                    at = quantile(diff_genes_matrix, c(<span class="number">0</span>,<span class="number">1</span>)), </span><br><span class="line">                                    labels = c(<span class="string">"Min"</span>, <span class="string">"Max"</span>),legend_height = unit(<span class="number">4</span>, <span class="string">"cm"</span>),</span><br><span class="line">                                    title_position = <span class="string">"leftcenter-rot"</span>)</span><br><span class="line">)</span><br></pre></td></tr></table></figure></li></ul><p><img src="https://showteeth.oss-cn-beijing.aliyuncs.com/blog_img/complexheatmap_order_row_and_split.png" alt="complexheatmap_order_row_and_split.png"></p><hr><h2 id="设置rowname的左右"><a href="#设置rowname的左右" class="headerlink" title="设置rowname的左右"></a>设置rowname的左右</h2><p><strong>默认情况下rowname是在右边，而聚类是在左边</strong>，如果想修改这顺序可以使用：<br></p><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">row_names_side = <span class="string">"left"</span>, row_dend_side = <span class="string">"right"</span></span><br></pre></td></tr></table></figure><p></p><hr><h2 id="设置cell的大小"><a href="#设置cell的大小" class="headerlink" title="设置cell的大小"></a>设置cell的大小</h2><p><code>ComplexHeatmap</code>没有直接设置cell大小的参数，但是可以通过设置整个heatmap的大小来达到目的：<br></p><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">Heatmap(mat, <span class="keyword">...</span>, heatmap_height = unit(<span class="number">1</span>, <span class="string">"cm"</span>)*nrow(mat))</span><br></pre></td></tr></table></figure><p></p><p><code>width</code> ， <code>heatmap_width</code> ， <code>height</code> 和 <code>heatmap_height</code> 都用于<strong>控制热图的大小</strong>。默认情况下，<strong>所有热图组件都具有固定的宽度或高度，例如 行树形图的宽度为 1cm</strong> 。<code>heatmap_width</code> 和 <code>heatmap_height</code> 控制<strong>整个热图的宽度/高度</strong>，<strong>包括所有热图组件（不包括图例）</strong>，而 <code>width</code> 和 <code>height</code> 仅控制 <strong>heamtap 主体的宽度/高度</strong>.</p><hr><h2 id="对某些行名和列名进行注释"><a href="#对某些行名和列名进行注释" class="headerlink" title="对某些行名和列名进行注释"></a>对某些行名和列名进行注释</h2><p>有时候显示全部的行名或者列名可能会比较杂乱，特别是行或者列很多的情况下，这个时候可以选择一些比较重要的进行展示：<br></p><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 在Heatmap中添加如下</span></span><br><span class="line"><span class="comment"># link_height:控制线的长短</span></span><br><span class="line">labels = <span class="string">"Spp1"</span></span><br><span class="line">top_annotation = columnAnnotation(link = anno_mark(at = c(<span class="number">33</span>), </span><br><span class="line">                                                   labels = labels,</span><br><span class="line">                                                   link_height = unit(<span class="number">4</span>, <span class="string">"mm"</span>))</span><br><span class="line">                                  )</span><br></pre></td></tr></table></figure><p></p><p>注意：这个<strong>只能放在<code>top_annotation</code></strong>，如果是<code>bottom_annotation</code>会发现那个连接线会出现在下方，而不是个热图连接起来，其实还是比较适合<strong>在行上进行注释，会比较好看一些</strong>。</p><p><img src="https://showteeth.oss-cn-beijing.aliyuncs.com/blog_img/complexheatmap_anno_col_row_names.png" alt="complexheatmap_anno_col_row_names.png"></p><hr><h2 id="在行或者列添加注释块"><a href="#在行或者列添加注释块" class="headerlink" title="在行或者列添加注释块"></a>在行或者列添加注释块</h2><p>在做热图中不可避免地会遇到需要<strong>将多个重复或者一类的样本进行标记的需求</strong>，这里就学习在行或者列上添加这种<strong>注释块</strong>：<br></p><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 先得到注释块的大小和颜色等信息</span></span><br><span class="line"><span class="comment"># 这里的a、b、c会显示成legend的名称</span></span><br><span class="line"><span class="comment"># gap：注释块与热图之间的距离</span></span><br><span class="line">df = data.frame(type = c(rep(<span class="string">"a"</span>, <span class="number">3</span>), rep(<span class="string">"b"</span>, <span class="number">3</span>), rep(<span class="string">"c"</span>, <span class="number">3</span>)))</span><br><span class="line">ha = rowAnnotation(df = df, col = list(type = c(<span class="string">"a"</span> = <span class="string">"red"</span>, <span class="string">"b"</span> = <span class="string">"blue"</span>,<span class="string">"c"</span>=<span class="string">"black"</span>)),show_legend = <span class="literal">F</span>,</span><br><span class="line">                   width = unit(<span class="number">1</span>, <span class="string">"cm"</span>),show_annotation_name = <span class="literal">F</span>, gap = unit(<span class="number">1</span>, <span class="string">"points"</span>))</span><br><span class="line"></span><br><span class="line"><span class="comment"># 在Heatmap中添加如下</span></span><br><span class="line">left_annotation = ha</span><br><span class="line"></span><br><span class="line"><span class="comment"># 完整版</span></span><br><span class="line">df = data.frame(type = c(rep(<span class="string">"a"</span>, <span class="number">3</span>), rep(<span class="string">"b"</span>, <span class="number">3</span>), rep(<span class="string">"c"</span>, <span class="number">3</span>)))</span><br><span class="line">ha = rowAnnotation(df = df, col = list(type = c(<span class="string">"a"</span> = <span class="string">"red"</span>, <span class="string">"b"</span> = <span class="string">"blue"</span>,<span class="string">"c"</span>=<span class="string">"black"</span>)),show_legend = <span class="literal">F</span>,</span><br><span class="line">                   width = unit(<span class="number">1</span>, <span class="string">"cm"</span>),show_annotation_name = <span class="literal">F</span>,gap = unit(<span class="number">1</span>, <span class="string">"points"</span>))</span><br><span class="line"></span><br><span class="line">Heatmap(diff_genes_matrix, show_column_names = <span class="literal">FALSE</span>, </span><br><span class="line">        show_row_dend = <span class="literal">FALSE</span>, show_column_dend = <span class="literal">F</span>, </span><br><span class="line">        col=red_green,</span><br><span class="line">        name=<span class="string">"Expression level"</span>,</span><br><span class="line">        row_order = sort(rownames(diff_genes_matrix)),</span><br><span class="line">        row_names_side = <span class="string">"left"</span>,</span><br><span class="line">        heatmap_width = unit(<span class="number">24</span>, <span class="string">"cm"</span>), heatmap_height = unit(<span class="number">8</span>, <span class="string">"cm"</span>),</span><br><span class="line">        top_annotation = columnAnnotation(link = anno_mark(at = c(<span class="number">33</span>), </span><br><span class="line">                                                           labels = labels,</span><br><span class="line">                                                           link_height = unit(<span class="number">4</span>, <span class="string">"mm"</span>)</span><br><span class="line">                                                           )</span><br><span class="line">                                          ),</span><br><span class="line">        left_annotation = ha,</span><br><span class="line">        heatmap_legend_param = list(color_bar = <span class="string">"continuous"</span>, </span><br><span class="line">                                    at = quantile(diff_genes_matrix, c(<span class="number">0</span>,<span class="number">1</span>)), </span><br><span class="line">                                    labels = c(<span class="string">"Min"</span>, <span class="string">"Max"</span>),legend_height = unit(<span class="number">4</span>, <span class="string">"cm"</span>),</span><br><span class="line">                                    title_position = <span class="string">"leftcenter-rot"</span>)</span><br><span class="line">        )</span><br></pre></td></tr></table></figure><p></p><p><img src="https://showteeth.oss-cn-beijing.aliyuncs.com/blog_img/complexheatmap_anno_block.png" alt="complexheatmap_anno_block.png"></p><hr><h2 id="将不同的注释块分开"><a href="#将不同的注释块分开" class="headerlink" title="将不同的注释块分开"></a>将不同的注释块分开</h2><p>在上一步得到不同的注释块了之后可以进一步地将不同的注释块之间分开：<br></p><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 设置分开的dataframe</span></span><br><span class="line">split = data.frame(rep(c(<span class="string">"A"</span>, <span class="string">"B"</span>,<span class="string">"C"</span>), each=<span class="number">3</span>))</span><br><span class="line"></span><br><span class="line"><span class="comment"># Heatmap中相关设置</span></span><br><span class="line"><span class="comment">## 设置分隔</span></span><br><span class="line">row_split = split,</span><br><span class="line"><span class="comment">## 去除分隔后添加的title：A、B、C</span></span><br><span class="line">row_title = <span class="literal">NULL</span>,</span><br><span class="line"><span class="comment">## 设置分隔的间隔大小</span></span><br><span class="line">row_gap = unit(<span class="number">5</span>, <span class="string">"mm"</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 完整设置</span></span><br><span class="line">split = data.frame(rep(c(<span class="string">"A"</span>, <span class="string">"B"</span>,<span class="string">"C"</span>), each=<span class="number">3</span>))</span><br><span class="line">Heatmap(diff_genes_matrix, show_column_names = <span class="literal">FALSE</span>, </span><br><span class="line">        show_row_dend = <span class="literal">FALSE</span>, show_column_dend = <span class="literal">F</span>, </span><br><span class="line">        col=red_green,</span><br><span class="line">        name=<span class="string">"Expression level"</span>,</span><br><span class="line">        row_order = sort(rownames(diff_genes_matrix)),</span><br><span class="line">        row_names_side = <span class="string">"left"</span>,</span><br><span class="line">        heatmap_width = unit(<span class="number">24</span>, <span class="string">"cm"</span>), heatmap_height = unit(<span class="number">8</span>, <span class="string">"cm"</span>),</span><br><span class="line">        top_annotation = columnAnnotation(link = anno_mark(at = c(<span class="number">33</span>), </span><br><span class="line">                                                           labels = labels,</span><br><span class="line">                                                           link_height = unit(<span class="number">4</span>, <span class="string">"mm"</span>)</span><br><span class="line">                                                           )</span><br><span class="line">                                          ),</span><br><span class="line">        left_annotation = ha,</span><br><span class="line">        row_split = split,</span><br><span class="line">        row_title = <span class="literal">NULL</span>,</span><br><span class="line">        row_gap = unit(<span class="number">5</span>, <span class="string">"mm"</span>),</span><br><span class="line">        heatmap_legend_param = list(color_bar = <span class="string">"continuous"</span>, </span><br><span class="line">                                    at = quantile(diff_genes_matrix, c(<span class="number">0</span>,<span class="number">1</span>)), </span><br><span class="line">                                    labels = c(<span class="string">"Min"</span>, <span class="string">"Max"</span>),legend_height = unit(<span class="number">4</span>, <span class="string">"cm"</span>),</span><br><span class="line">                                    title_position = <span class="string">"leftcenter-rot"</span>)</span><br><span class="line">        )</span><br></pre></td></tr></table></figure><p></p><p><img src="https://showteeth.oss-cn-beijing.aliyuncs.com/blog_img/complexheatmap_anno_block_split.png" alt="complexheatmap_anno_block_split.png"></p><hr><h2 id="完整脚本"><a href="#完整脚本" class="headerlink" title="完整脚本"></a>完整脚本</h2><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">library</span>(ComplexHeatmap)</span><br><span class="line"><span class="keyword">library</span>(circlize)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 对counts做了变换，log2(normalization + 1)</span></span><br><span class="line">diff_genes_matrix=t(as.matrix(log2(diff_genes_df+<span class="number">1</span>)))</span><br><span class="line">diff_genes_matrix</span><br><span class="line">            Cdh4   Sema6b    Meox1   Col1a1     Ell2    Tcea3   Arrdc2    Casq1    Bcl2l1   Piezo1   Mov10l1</span><br><span class="line">  BE_W_1 <span class="number">5.674636</span> <span class="number">8.992203</span> <span class="number">8.271125</span> <span class="number">10.80915</span> <span class="number">9.010202</span> <span class="number">10.84969</span> <span class="number">8.550702</span> <span class="number">6.719983</span> <span class="number">10.534402</span> <span class="number">9.251238</span>  <span class="number">9.815931</span></span><br><span class="line">  BE_W_3 <span class="number">6.278332</span> <span class="number">8.937379</span> <span class="number">7.755463</span> <span class="number">10.48085</span> <span class="number">8.901625</span> <span class="number">10.94128</span> <span class="number">8.931936</span> <span class="number">6.613638</span> <span class="number">10.654491</span> <span class="number">9.340083</span> <span class="number">10.054614</span></span><br><span class="line">  BE_W_6 <span class="number">6.053069</span> <span class="number">9.046550</span> <span class="number">8.143926</span> <span class="number">10.77555</span> <span class="number">9.108047</span> <span class="number">10.90593</span> <span class="number">9.411026</span> <span class="number">6.555065</span> <span class="number">10.923830</span> <span class="number">9.369440</span> <span class="number">10.122322</span></span><br><span class="line">  BE_C_2 <span class="number">6.970798</span> <span class="number">8.712448</span> <span class="number">7.825648</span> <span class="number">10.61191</span> <span class="number">8.854695</span> <span class="number">10.90589</span> <span class="number">8.221351</span> <span class="number">6.415780</span> <span class="number">10.006843</span> <span class="number">9.411790</span>  <span class="number">9.938315</span></span><br><span class="line">  ......</span><br><span class="line"></span><br><span class="line">labels = <span class="string">"Spp1"</span></span><br><span class="line"></span><br><span class="line">red_green=colorRamp2(breaks=c(min(diff_genes_matrix), </span><br><span class="line">                       mean(diff_genes_matrix), </span><br><span class="line">                       max(diff_genes_matrix)),</span><br><span class="line">                       colors=c(<span class="string">"red"</span>, <span class="string">"black"</span>, <span class="string">"green"</span>))</span><br><span class="line"></span><br><span class="line">df = data.frame(type = c(rep(<span class="string">"a"</span>, <span class="number">3</span>), rep(<span class="string">"b"</span>, <span class="number">3</span>), rep(<span class="string">"c"</span>, <span class="number">3</span>)))</span><br><span class="line">ha = rowAnnotation(df = df, col = list(type = c(<span class="string">"a"</span> = <span class="string">"red"</span>, <span class="string">"b"</span> = <span class="string">"blue"</span>,<span class="string">"c"</span>=<span class="string">"black"</span>)),show_legend = <span class="literal">F</span>,</span><br><span class="line">                   width = unit(<span class="number">1</span>, <span class="string">"cm"</span>),show_annotation_name = <span class="literal">F</span>,gap = unit(<span class="number">1</span>, <span class="string">"points"</span>))</span><br><span class="line"></span><br><span class="line">split = data.frame(rep(c(<span class="string">"A"</span>, <span class="string">"B"</span>,<span class="string">"C"</span>), each=<span class="number">3</span>))</span><br><span class="line"></span><br><span class="line">Heatmap(diff_genes_matrix, show_column_names = <span class="literal">FALSE</span>, </span><br><span class="line">        show_row_dend = <span class="literal">FALSE</span>, show_column_dend = <span class="literal">F</span>, </span><br><span class="line">        col=red_green,</span><br><span class="line">        name=<span class="string">"Expression level"</span>,</span><br><span class="line">        row_order = sort(rownames(diff_genes_matrix)),</span><br><span class="line">        row_names_side = <span class="string">"left"</span>,</span><br><span class="line">        heatmap_width = unit(<span class="number">24</span>, <span class="string">"cm"</span>), heatmap_height = unit(<span class="number">8</span>, <span class="string">"cm"</span>),</span><br><span class="line">        top_annotation = columnAnnotation(link = anno_mark(at = c(<span class="number">33</span>), </span><br><span class="line">                                                           labels = labels,</span><br><span class="line">                                                           link_height = unit(<span class="number">4</span>, <span class="string">"mm"</span>)</span><br><span class="line">                                                           )</span><br><span class="line">                                          ),</span><br><span class="line">        left_annotation = ha,</span><br><span class="line">        row_split = split,</span><br><span class="line">        row_title = <span class="literal">NULL</span>,</span><br><span class="line">        row_gap = unit(<span class="number">2</span>, <span class="string">"mm"</span>),</span><br><span class="line">        heatmap_legend_param = list(color_bar = <span class="string">"continuous"</span>, </span><br><span class="line">                                    at = quantile(diff_genes_matrix, c(<span class="number">0</span>,<span class="number">1</span>)), </span><br><span class="line">                                    labels = c(<span class="string">"Min"</span>, <span class="string">"Max"</span>),legend_height = unit(<span class="number">4</span>, <span class="string">"cm"</span>),</span><br><span class="line">                                    title_position = <span class="string">"leftcenter-rot"</span>)</span><br><span class="line">        )</span><br></pre></td></tr></table></figure><p>最终版的图见上！</p><hr><h2 id="oncoPrint-适合展示突变"><a href="#oncoPrint-适合展示突变" class="headerlink" title="oncoPrint-适合展示突变"></a>oncoPrint-适合展示突变</h2><p><code>ComplexHeatmap</code>专门做了个<a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/oncoprint.html#" target="_blank" rel="noopener">oncoPrint函数</a>来<strong>绘制突变相关信息</strong>，感觉和<a href="https://github.com/PoisonAlien/maftools" target="_blank" rel="noopener">maftools</a>的作用很类似，后续可以重点学习和使用一下。</p><h3 id="输入数据"><a href="#输入数据" class="headerlink" title="输入数据"></a>输入数据</h3><h4 id="使用分隔符分割的多种突变名称"><a href="#使用分隔符分割的多种突变名称" class="headerlink" title="使用分隔符分割的多种突变名称"></a>使用分隔符分割的多种突变名称</h4><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 示例数据</span></span><br><span class="line">mat = read.table(textConnection(</span><br><span class="line">  <span class="string">"s1,s2,s3</span></span><br><span class="line"><span class="string">g1,snv;indel,snv,indel</span></span><br><span class="line"><span class="string">g2,,snv;indel,snv</span></span><br><span class="line"><span class="string">g3,snv,,indel;snv"</span>), row.names = <span class="number">1</span>, header = <span class="literal">TRUE</span>, sep = <span class="string">","</span>, stringsAsFactors = <span class="literal">FALSE</span>)</span><br><span class="line">mat = as.matrix(mat)</span><br><span class="line">mat</span><br><span class="line">                s1          s2          s3         </span><br><span class="line">        g1 <span class="string">"snv;indel"</span> <span class="string">"snv"</span>       <span class="string">"indel"</span>    </span><br><span class="line">        g2 <span class="string">""</span>          <span class="string">"snv;indel"</span> <span class="string">"snv"</span>      </span><br><span class="line">        g3 <span class="string">"snv"</span>       <span class="string">""</span>          <span class="string">"indel;snv"</span></span><br></pre></td></tr></table></figure><hr><h4 id="使用0-1表示的多个variant列表"><a href="#使用0-1表示的多个variant列表" class="headerlink" title="使用0/1表示的多个variant列表"></a>使用0/1表示的多个variant列表</h4><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">mat_list = list(snv = matrix(c(<span class="number">1</span>, <span class="number">0</span>, <span class="number">1</span>, <span class="number">1</span>, <span class="number">1</span>, <span class="number">0</span>, <span class="number">0</span>, <span class="number">1</span>, <span class="number">1</span>), nrow = <span class="number">3</span>),</span><br><span class="line">                indel = matrix(c(<span class="number">1</span>, <span class="number">0</span>, <span class="number">0</span>, <span class="number">0</span>, <span class="number">1</span>, <span class="number">0</span>, <span class="number">1</span>, <span class="number">0</span>, <span class="number">0</span>), nrow = <span class="number">3</span>))</span><br><span class="line">rownames(mat_list$snv) = rownames(mat_list$indel) = c(<span class="string">"g1"</span>, <span class="string">"g2"</span>, <span class="string">"g3"</span>)</span><br><span class="line">colnames(mat_list$snv) = colnames(mat_list$indel) = c(<span class="string">"s1"</span>, <span class="string">"s2"</span>, <span class="string">"s3"</span>)</span><br><span class="line">mat_list</span><br><span class="line">        $snv</span><br><span class="line">           s1 s2 s3</span><br><span class="line">        g1  <span class="number">1</span>  <span class="number">1</span>  <span class="number">0</span></span><br><span class="line">        g2  <span class="number">0</span>  <span class="number">1</span>  <span class="number">1</span></span><br><span class="line">        g3  <span class="number">1</span>  <span class="number">0</span>  <span class="number">1</span></span><br><span class="line"></span><br><span class="line">        $indel</span><br><span class="line">           s1 s2 s3</span><br><span class="line">        g1  <span class="number">1</span>  <span class="number">0</span>  <span class="number">1</span></span><br><span class="line">        g2  <span class="number">0</span>  <span class="number">1</span>  <span class="number">0</span></span><br><span class="line">        g3  <span class="number">0</span>  <span class="number">0</span>  <span class="number">0</span></span><br></pre></td></tr></table></figure><div class="note info"><p><code>oncoPrint()</code>要求<code>mat_list</code>中的<strong>所有元素都有相同的行名和列名</strong>，并且绘图也不用添加<code>get_type</code>参数。</p></div><hr><h3 id="得到突变类别"><a href="#得到突变类别" class="headerlink" title="得到突变类别"></a>得到突变类别</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 示例数据中一个基因可能同时存在多种突变类别，多种突变类别是使用;分割的</span></span><br><span class="line"><span class="comment"># 得到突变类别的方法</span></span><br><span class="line">get_type_fun = <span class="keyword">function</span>(x) strsplit(x, <span class="string">";"</span>)[[<span class="number">1</span>]]</span><br><span class="line">get_type_fun(mat[<span class="number">1</span>, <span class="number">1</span>])</span><br><span class="line">        [<span class="number">1</span>] <span class="string">"snv"</span>   <span class="string">"indel"</span></span><br></pre></td></tr></table></figure><div class="note info"><p>针对常见的一些分隔符，如<code>;</code>、<code>:</code>、<code>,</code>、<code>|</code>，<code>oncoPrint()</code>都会自动的将其分割，不用单独指定；如果自己数据中的分隔符不在常见分隔符中，需要自定义<code>get_type_fun</code>，并将其传递给<code>oncoPrint()</code>的<code>get_type</code>参数。</p></div><hr><h3 id="设置variant的颜色"><a href="#设置variant的颜色" class="headerlink" title="设置variant的颜色"></a>设置variant的颜色</h3><p>不同类型variant的颜色可以使用<code>oncoPrint()</code>的col参数进行设置，col参数是一个带名称的向量：<br></p><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 带名称的向量</span></span><br><span class="line">col = c(snv = <span class="string">"red"</span>, indel = <span class="string">"blue"</span>)</span><br><span class="line">col</span><br><span class="line">        snv  indel </span><br><span class="line">        <span class="string">"red"</span> <span class="string">"blue"</span></span><br></pre></td></tr></table></figure><p></p><hr><h3 id="设置variant的形状"><a href="#设置variant的形状" class="headerlink" title="设置variant的形状"></a>设置variant的形状</h3><p>一个样本的一个基因可能会存在多种突变形式，在画图的时候如果不做区别就会<strong>出现重叠</strong>的现象，<code>oncoPrint()</code>提供了<code>alter_fun</code>参数来修改不同突变的展示形状，便于区分。</p><p><code>alter_fun</code>可以是一个包含了逐层添加图形的函数列表(比如：先画snv的图层，接着画indel的图层)，这种情况下<code>alter_fun</code>中的函数应该包含四个参数：</p><ul><li><code>x</code>和<code>y</code>表示坐标位置：positions of the grids on the oncoPrint (x and y)</li><li><code>w</code>和<code>h</code>表示宽度和高度：widths and heights of the grids (w and h, which is measured in <code>npc unit</code>)</li></ul><p>同时，<code>alter_fun</code>也可以通过一个单一的函数来指定，这种情况下<code>alter_fun</code>不再是一个函数的列表，而是一个<strong>包含了五个参数的函数</strong>，增加的参数是<code>v</code>，这个参数是一个<strong>逻辑值</strong>，用于<strong>判断某个样本中的某个基因是不是包含了某个突变</strong>，其也是一个带名称的向量：<br></p><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 某个样本的某个基因的v参数示例</span></span><br><span class="line"><span class="comment"># 只包括一个snv</span></span><br><span class="line">snv indel </span><br><span class="line"><span class="literal">TRUE</span> <span class="literal">FALSE</span></span><br></pre></td></tr></table></figure><p></p><p>使用<strong>函数列表</strong>以及<strong>使用单个函数</strong>的示例：<br></p><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 使用函数列表，包含四个参数</span></span><br><span class="line">oncoPrint(mat,</span><br><span class="line">    alter_fun = list(</span><br><span class="line">        snv = <span class="keyword">function</span>(x, y, w, h) grid.rect(x, y, w*<span class="number">0.9</span>, h*<span class="number">0.9</span>, </span><br><span class="line">            gp = gpar(fill = col[<span class="string">"snv"</span>], col = <span class="literal">NA</span>)),</span><br><span class="line">        indel = <span class="keyword">function</span>(x, y, w, h) grid.rect(x, y, w*<span class="number">0.9</span>, h*<span class="number">0.4</span>, </span><br><span class="line">            gp = gpar(fill = col[<span class="string">"indel"</span>], col = <span class="literal">NA</span>))</span><br><span class="line">    ), col = col)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 使用单个函数，添加了v作为逻辑值来判断</span></span><br><span class="line">oncoPrint(mat,</span><br><span class="line">    alter_fun = <span class="keyword">function</span>(x, y, w, h, v) &#123;</span><br><span class="line">        <span class="keyword">if</span>(v[<span class="string">"snv"</span>]) grid.rect(x, y, w*<span class="number">0.9</span>, h*<span class="number">0.9</span>, <span class="comment"># v["snv"] is a logical value</span></span><br><span class="line">            gp = gpar(fill = col[<span class="string">"snv"</span>], col = <span class="literal">NA</span>))</span><br><span class="line">        <span class="keyword">if</span>(v[<span class="string">"indel"</span>]) grid.rect(x, y, w*<span class="number">0.9</span>, h*<span class="number">0.4</span>, <span class="comment"># v["indel"] is a logical value</span></span><br><span class="line">            gp = gpar(fill = col[<span class="string">"indel"</span>], col = <span class="literal">NA</span>))</span><br><span class="line">    &#125;, col = col)</span><br></pre></td></tr></table></figure><p></p><h4 id="关于grid的位置参数"><a href="#关于grid的位置参数" class="headerlink" title="关于grid的位置参数"></a>关于grid的位置参数</h4><p>因为绘图需要涉及到x、y、w、h四个grid绘图语法的参数，因为自身对grid语法并不是很了解，所以这里补充一下：<br></p><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">library</span>(grid)</span><br><span class="line">grid.show.viewport(viewport(</span><br><span class="line">  x = <span class="number">0.6</span>, y = <span class="number">0.6</span>,</span><br><span class="line">  w = unit(<span class="number">1</span>, <span class="string">"inches"</span>), h = unit(<span class="number">1</span>, <span class="string">"inches"</span>)</span><br><span class="line">))</span><br></pre></td></tr></table></figure><p></p><p><img src="https://showteeth.oss-cn-beijing.aliyuncs.com/blog_img/complexheatmap_oncoprint_grid.png" alt="complexheatmap_oncoprint_grid.png"></p><p>从图形中可以看出，<code>(x,y)</code>其实是单个grid的中心点，其实这个主要是由参数<code>just</code>控制的，默认<code>just</code>为<code>centre</code>表明<code>(x,y)</code>位于grid的中心位置。</p><p>修改参数<code>just</code>值可以很明显看出来区别：<br></p><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 设置(x,y)位于top，也就是上中</span></span><br><span class="line">grid.show.viewport(viewport(</span><br><span class="line">  x = <span class="number">0.6</span>, y = <span class="number">0.6</span>,</span><br><span class="line">  w = unit(<span class="number">1</span>, <span class="string">"inches"</span>), h = unit(<span class="number">1</span>, <span class="string">"inches"</span>),</span><br><span class="line">  just=<span class="string">"top"</span></span><br><span class="line">))</span><br></pre></td></tr></table></figure><p></p><p><img src="https://showteeth.oss-cn-beijing.aliyuncs.com/blog_img/complexheatmap_oncoprint_grid_2.png" alt="complexheatmap_oncoprint_grid_2.png"></p><hr><h4 id="按比例绘图"><a href="#按比例绘图" class="headerlink" title="按比例绘图"></a>按比例绘图</h4><p>使用单个的函数来绘制图形可以更方便的定制化：<br></p><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 按比例画图</span></span><br><span class="line"><span class="comment"># 可以直接将mat替换为前面出现过的mat_list</span></span><br><span class="line">oncoPrint(mat,</span><br><span class="line">          alter_fun = <span class="keyword">function</span>(x, y, w, h, v) &#123;</span><br><span class="line">            n = sum(v)  <span class="comment"># how many alterations for current gene in current sample</span></span><br><span class="line">            h = h*<span class="number">0.9</span></span><br><span class="line">            <span class="comment"># use `names(which(v))` to correctly map between `v` and `col`</span></span><br><span class="line">            <span class="keyword">if</span>(n) grid.rect(x, y - h*<span class="number">0.5</span> + <span class="number">1</span>:n/n*h, w*<span class="number">0.9</span>, <span class="number">1</span>/n*h, </span><br><span class="line">                            gp = gpar(fill = col[names(which(v))], col = <span class="literal">NA</span>), just = <span class="string">"top"</span>)</span><br><span class="line">          &#125;, col = col)</span><br></pre></td></tr></table></figure><p></p><p>解释一下：</p><ul><li><code>n = sum(v)</code>：某个样本在某个基因上的突变数目(不进行类别上的区分)</li><li><code>h = h*0.9</code>：设置的是每个图形显示的大小，后面的那个h应该是某个grid的完整大小，0.9倍用于分隔不同的基因。</li><li><strong>最后的if语句</strong>：假定是一个存在一个snv和一个indel的基因，此时的<code>n=2</code>，另个小grid的坐标分别是<code>(x,y)</code>、<code>(x,y+h/2)</code>，需要注意的是这里的<code>just</code>参数的值是<code>top</code>，也就是这里的坐标都是<strong>上中的坐标</strong>。</li></ul><p><img src="https://showteeth.oss-cn-beijing.aliyuncs.com/blog_img/complexheatmap_oncoprint_pro_1.png" alt="complexheatmap_oncoprint_pro_1.png"></p><hr><h4 id="绘制三角形图案"><a href="#绘制三角形图案" class="headerlink" title="绘制三角形图案"></a>绘制三角形图案</h4><p>三角形图案的优点在于<strong>更加直观</strong>相比于成比例的块状图案，但是其<strong>表征的variant种类会更加少一些</strong>，因为一个grid分为两个三角形(更复杂的图案就比较难了)，而块状图案可以按比例分为很多小块；并且<strong>使用的话也就更加复杂一些</strong>，首先是要使用到<code>grid.polygon</code>这个函数，其主要是用于绘制多边形，用其来绘制三角形图案的时候需要提供的<code>x</code>和<code>y</code>都是<strong>三个参数的向量</strong>，所以比较复杂。<br></p><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line">oncoPrint(mat,</span><br><span class="line">    alter_fun = list(</span><br><span class="line">        background = <span class="keyword">function</span>(x, y, w, h) &#123;</span><br><span class="line">            grid.polygon(</span><br><span class="line">                <span class="comment"># unit.c(x - 0.5*w, x - 0.5*w, x + 0.5*w)是一组x的坐标，其中每个与后面的unit.c(y - 0.5*h, y + 0.5*h, y - 0.5*h)一一对应</span></span><br><span class="line">                unit.c(x - <span class="number">0.5</span>*w, x - <span class="number">0.5</span>*w, x + <span class="number">0.5</span>*w), </span><br><span class="line">                unit.c(y - <span class="number">0.5</span>*h, y + <span class="number">0.5</span>*h, y - <span class="number">0.5</span>*h),</span><br><span class="line">                gp = gpar(fill = <span class="string">"grey"</span>, col = <span class="string">"white"</span>))</span><br><span class="line">            grid.polygon(</span><br><span class="line">                unit.c(x + <span class="number">0.5</span>*w, x + <span class="number">0.5</span>*w, x - <span class="number">0.5</span>*w), </span><br><span class="line">                unit.c(y + <span class="number">0.5</span>*h, y - <span class="number">0.5</span>*h, y + <span class="number">0.5</span>*h),</span><br><span class="line">                gp = gpar(fill = <span class="string">"grey"</span>, col = <span class="string">"white"</span>))</span><br><span class="line">        &#125;,</span><br><span class="line">        snv = <span class="keyword">function</span>(x, y, w, h) &#123;</span><br><span class="line">            grid.polygon(</span><br><span class="line">                unit.c(x - <span class="number">0.5</span>*w, x - <span class="number">0.5</span>*w, x + <span class="number">0.5</span>*w), </span><br><span class="line">                unit.c(y - <span class="number">0.5</span>*h, y + <span class="number">0.5</span>*h, y - <span class="number">0.5</span>*h),</span><br><span class="line">                gp = gpar(fill = col[<span class="string">"snv"</span>], col = <span class="string">"white"</span>))</span><br><span class="line">        &#125;,</span><br><span class="line">        indel = <span class="keyword">function</span>(x, y, w, h) &#123;</span><br><span class="line">            grid.polygon(</span><br><span class="line">                unit.c(x + <span class="number">0.5</span>*w, x + <span class="number">0.5</span>*w, x - <span class="number">0.5</span>*w), </span><br><span class="line">                unit.c(y + <span class="number">0.5</span>*h, y - <span class="number">0.5</span>*h, y + <span class="number">0.5</span>*h),</span><br><span class="line">                gp = gpar(fill = col[<span class="string">"indel"</span>], col = <span class="string">"white"</span>))</span><br><span class="line">        &#125;</span><br><span class="line">    ), col = col)</span><br></pre></td></tr></table></figure><p></p><p><img src="https://showteeth.oss-cn-beijing.aliyuncs.com/blog_img/complexheatmap_oncoprint_tri.png" alt="complexheatmap_oncoprint_tri.png"></p><hr><h4 id="用于支持更多variant类型"><a href="#用于支持更多variant类型" class="headerlink" title="用于支持更多variant类型"></a>用于支持更多variant类型</h4><p>某个样本的某个基因出现多种突变类型是很常见的事，所以如何在单个grid中表征更多的突变类型就非常关键，<code>oncoPrint</code>针对这种情况设计了<code>test_alter_fun()</code>函数来查看设计的<code>alter_fun</code>形状。<br></p><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">alter_fun = list(</span><br><span class="line">    mut1 = <span class="keyword">function</span>(x, y, w, h) </span><br><span class="line">        grid.rect(x, y, w, h, gp = gpar(fill = <span class="string">"red"</span>, col = <span class="literal">NA</span>)),</span><br><span class="line">    mut2 = <span class="keyword">function</span>(x, y, w, h) </span><br><span class="line">        grid.rect(x, y, w, h, gp = gpar(fill = <span class="string">"blue"</span>, col = <span class="literal">NA</span>)),</span><br><span class="line">    mut3 = <span class="keyword">function</span>(x, y, w, h) </span><br><span class="line">        grid.rect(x, y, w, h, gp = gpar(fill = <span class="string">"yellow"</span>, col = <span class="literal">NA</span>)),</span><br><span class="line">    mut4 = <span class="keyword">function</span>(x, y, w, h) </span><br><span class="line">        grid.rect(x, y, w, h, gp = gpar(fill = <span class="string">"purple"</span>, col = <span class="literal">NA</span>)),</span><br><span class="line">    mut5 = <span class="keyword">function</span>(x, y, w, h) </span><br><span class="line">        grid.rect(x, y, w, h, gp = gpar(fill = <span class="literal">NA</span>, lwd = <span class="number">2</span>)),</span><br><span class="line">    mut6 = <span class="keyword">function</span>(x, y, w, h) </span><br><span class="line">        grid.points(x, y, pch = <span class="number">16</span>),</span><br><span class="line">    mut7 = <span class="keyword">function</span>(x, y, w, h) </span><br><span class="line">        grid.segments(x - w*<span class="number">0.5</span>, y - h*<span class="number">0.5</span>, x + w*<span class="number">0.5</span>, y + h*<span class="number">0.5</span>, gp = gpar(lwd = <span class="number">2</span>))</span><br><span class="line">)</span><br><span class="line">test_alter_fun(alter_fun)</span><br></pre></td></tr></table></figure><p></p><p>虽然下面这种给出了14中不同的组合类型，但是<strong>其实存在一些问题</strong>，比如<strong>前4中mut类型之间如果存在重叠的话就会出现相互遮挡的问题</strong>，比较好的解决方法就是使用前面的按比例分配。所以个人的想法是：<mark>纯色的用来表征不同的突变类型，然后剩下的几种框、点、线用于表征其他信息，比如表达值、CNV信息。</mark></p><p><img src="https://showteeth.oss-cn-beijing.aliyuncs.com/blog_img/complexheatmap_oncoprint_test_alter_1.png" alt="complexheatmap_oncoprint_test_alter_1.png"></p><hr><h3 id="设置grid背景颜色"><a href="#设置grid背景颜色" class="headerlink" title="设置grid背景颜色"></a>设置grid背景颜色</h3><p>在以函数列表形式设置<code>alter_fun</code>参数的时候，不同的图形是按顺序加上了，所以如果想给每个grid添加背景的话可以在函数列表中第一个就添加一个<code>background</code>函数用于设置grid的背景颜色。<br></p><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 设置背景颜色为light green，并且添加了框线</span></span><br><span class="line">oncoPrint(mat,</span><br><span class="line">    alter_fun = list(</span><br><span class="line">        background = <span class="keyword">function</span>(x, y, w, h) grid.rect(x, y, w, h, </span><br><span class="line">            gp = gpar(fill = <span class="string">"#00FF0020"</span>)),</span><br><span class="line">        snv = <span class="keyword">function</span>(x, y, w, h) grid.rect(x, y, w*<span class="number">0.9</span>, h*<span class="number">0.9</span>, </span><br><span class="line">            gp = gpar(fill = col[<span class="string">"snv"</span>], col = <span class="literal">NA</span>)),</span><br><span class="line">        indel = <span class="keyword">function</span>(x, y, w, h) grid.rect(x, y, w*<span class="number">0.9</span>, h*<span class="number">0.4</span>, </span><br><span class="line">            gp = gpar(fill = col[<span class="string">"indel"</span>], col = <span class="literal">NA</span>))</span><br><span class="line">    ), col = col)</span><br></pre></td></tr></table></figure><p></p><p><img src="https://showteeth.oss-cn-beijing.aliyuncs.com/blog_img/complexheatmap_oncoprint_bg_2.png" alt="complexheatmap_oncoprint_bg_2.png"></p><p>如果想去除任何的框线以及颜色背景，可以将<code>background</code>设置为<code>NULL</code>:<br></p><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">oncoPrint(mat,</span><br><span class="line">    alter_fun = list(</span><br><span class="line">        <span class="comment"># 设置背景为NULL</span></span><br><span class="line">        background = <span class="keyword">function</span>(<span class="keyword">...</span>) <span class="literal">NULL</span>,</span><br><span class="line">        snv = <span class="keyword">function</span>(x, y, w, h) grid.rect(x, y, w*<span class="number">0.9</span>, h*<span class="number">0.9</span>, </span><br><span class="line">            gp = gpar(fill = col[<span class="string">"snv"</span>], col = <span class="literal">NA</span>)),</span><br><span class="line">        indel = <span class="keyword">function</span>(x, y, w, h) grid.rect(x, y, w*<span class="number">0.9</span>, h*<span class="number">0.4</span>, </span><br><span class="line">            gp = gpar(fill = col[<span class="string">"indel"</span>], col = <span class="literal">NA</span>))</span><br><span class="line">    ), col = col)</span><br></pre></td></tr></table></figure><p></p><p><img src="https://showteeth.oss-cn-beijing.aliyuncs.com/blog_img/complexheatmap_oncoprint_bg_1.png" alt="complexheatmap_oncoprint_bg_1.png"></p><hr><h3 id="按重要性绘制图层"><a href="#按重要性绘制图层" class="headerlink" title="按重要性绘制图层"></a>按重要性绘制图层</h3><p>在绘图时有些信息是主要信息，有些信息可能是次要信息，但是都是必须要展示的图形，这种情况下其实就是前面<a href="#%e7%94%a8%e4%ba%8e%e6%94%af%e6%8c%81%e6%9b%b4%e5%a4%9avariant%e7%b1%bb%e5%9e%8b">用于支持更多variant类型</a>最后我提到的用纯色表示不同的突变类型，然后剩下的几种框、点、线用于表征其他信息，比如表达值、CNV信息。<br></p><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 生成虚拟数据</span></span><br><span class="line">set.seed(<span class="number">123</span>)</span><br><span class="line">x1 = sample(c(<span class="string">""</span>, <span class="string">"snv"</span>), <span class="number">100</span>, replace = <span class="literal">TRUE</span>, prob = c(<span class="number">8</span>, <span class="number">2</span>))</span><br><span class="line">x2 = sample(c(<span class="string">""</span>, <span class="string">"indel"</span>), <span class="number">100</span>, replace = <span class="literal">TRUE</span>, prob = c(<span class="number">8</span>, <span class="number">2</span>))</span><br><span class="line">x2[x1 == <span class="string">"snv"</span>] = <span class="string">""</span></span><br><span class="line">x3 = sample(c(<span class="string">""</span>, <span class="string">"intronic"</span>), <span class="number">100</span>, replace = <span class="literal">TRUE</span>, prob = c(<span class="number">5</span>, <span class="number">5</span>))</span><br><span class="line">x4 = sample(c(<span class="string">""</span>, <span class="string">"exonic"</span>), <span class="number">100</span>, replace = <span class="literal">TRUE</span>, prob = c(<span class="number">5</span>, <span class="number">5</span>))</span><br><span class="line">x3[x1 == <span class="string">""</span> &amp; x2 == <span class="string">""</span>] = <span class="string">""</span></span><br><span class="line">x4[x1 == <span class="string">""</span> &amp; x2 == <span class="string">""</span>] = <span class="string">""</span></span><br><span class="line">x4[x3 == <span class="string">"intronic"</span>] = <span class="string">""</span></span><br><span class="line">x = apply(cbind(x1, x2, x3, x4), <span class="number">1</span>, <span class="keyword">function</span>(x) &#123;</span><br><span class="line">    x = x[x != <span class="string">""</span>]</span><br><span class="line">    paste(x, collapse = <span class="string">";"</span>)</span><br><span class="line">&#125;)</span><br><span class="line">m = matrix(x, nrow = <span class="number">10</span>, ncol = <span class="number">10</span>, dimnames = list(paste0(<span class="string">"g"</span>, <span class="number">1</span>:<span class="number">10</span>), paste0(<span class="string">"s"</span>, <span class="number">1</span>:<span class="number">10</span>)))</span><br><span class="line">m[<span class="number">1</span>:<span class="number">4</span>, <span class="number">1</span>:<span class="number">4</span>]</span><br><span class="line">           s1    s2             s3             s4            </span><br><span class="line">        g1 <span class="string">""</span>    <span class="string">"snv;intronic"</span> <span class="string">"snv;intronic"</span> <span class="string">"snv"</span>         </span><br><span class="line">        g2 <span class="string">""</span>    <span class="string">""</span>             <span class="string">""</span>             <span class="string">"snv;intronic"</span></span><br><span class="line">        g3 <span class="string">""</span>    <span class="string">""</span>             <span class="string">""</span>             <span class="string">""</span>            </span><br><span class="line">        g4 <span class="string">"snv"</span> <span class="string">"indel;exonic"</span> <span class="string">"snv"</span>          <span class="string">""</span>      </span><br><span class="line"></span><br><span class="line"><span class="comment"># 设置alter_fun</span></span><br><span class="line">alter_fun = list(</span><br><span class="line">    background = <span class="keyword">function</span>(x, y, w, h) </span><br><span class="line">        grid.rect(x, y, w*<span class="number">0.9</span>, h*<span class="number">0.9</span>, gp = gpar(fill = <span class="string">"#CCCCCC"</span>, col = <span class="literal">NA</span>)),</span><br><span class="line">    <span class="comment"># red rectangles</span></span><br><span class="line">    snv = <span class="keyword">function</span>(x, y, w, h) </span><br><span class="line">        grid.rect(x, y, w*<span class="number">0.9</span>, h*<span class="number">0.9</span>, gp = gpar(fill = <span class="string">"red"</span>, col = <span class="literal">NA</span>)),</span><br><span class="line">    <span class="comment"># blue rectangles</span></span><br><span class="line">    indel = <span class="keyword">function</span>(x, y, w, h) </span><br><span class="line">        grid.rect(x, y, w*<span class="number">0.9</span>, h*<span class="number">0.9</span>, gp = gpar(fill = <span class="string">"blue"</span>, col = <span class="literal">NA</span>)),</span><br><span class="line">    <span class="comment"># dots</span></span><br><span class="line">    intronic = <span class="keyword">function</span>(x, y, w, h) </span><br><span class="line">        grid.points(x, y, pch = <span class="number">16</span>),</span><br><span class="line">    <span class="comment"># crossed lines</span></span><br><span class="line">    exonic = <span class="keyword">function</span>(x, y, w, h) &#123;</span><br><span class="line">        grid.segments(x - w*<span class="number">0.4</span>, y - h*<span class="number">0.4</span>, x + w*<span class="number">0.4</span>, y + h*<span class="number">0.4</span>, gp = gpar(lwd = <span class="number">2</span>))</span><br><span class="line">        grid.segments(x + w*<span class="number">0.4</span>, y - h*<span class="number">0.4</span>, x - w*<span class="number">0.4</span>, y + h*<span class="number">0.4</span>, gp = gpar(lwd = <span class="number">2</span>))</span><br><span class="line">    &#125;</span><br><span class="line">)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 最终绘图</span></span><br><span class="line"><span class="comment"># we only define color for snv and indel, so barplot annotations only show snv and indel</span></span><br><span class="line">ht = oncoPrint(m, alter_fun = alter_fun, col = c(snv = <span class="string">"red"</span>, indel = <span class="string">"blue"</span>))</span><br><span class="line">draw(ht, heatmap_legend_list = list(</span><br><span class="line">    Legend(labels = c(<span class="string">"intronic"</span>, <span class="string">"exonic"</span>), type = <span class="string">"points"</span>, pch = c(<span class="number">16</span>, <span class="number">28</span>))</span><br><span class="line">))</span><br></pre></td></tr></table></figure><p></p><p><img src="https://showteeth.oss-cn-beijing.aliyuncs.com/blog_img/complexheatmap_oncoprint_intron_exon_1.png" alt="complexheatmap_oncoprint_intron_exon_1.png"></p><p>在单个函数中实现上述类似的用法：<br></p><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br></pre></td><td class="code"><pre><span class="line">oncoPrint(mat_list,</span><br><span class="line">          alter_fun = <span class="keyword">function</span>(x, y, w, h, v) &#123;</span><br><span class="line">            n = sum(v)  <span class="comment"># how many alterations for current gene in current sample</span></span><br><span class="line">            h = h*<span class="number">0.9</span></span><br><span class="line">            <span class="comment"># use `names(which(v))` to correctly map between `v` and `col`</span></span><br><span class="line">            <span class="keyword">if</span>(n) grid.rect(x, y - h*<span class="number">0.5</span> + <span class="number">1</span>:n/n*h, w*<span class="number">0.9</span>, <span class="number">1</span>/n*h, </span><br><span class="line">                            gp = gpar(fill = col[names(which(v))], col = <span class="literal">NA</span>), just = <span class="string">"top"</span>)</span><br><span class="line">            <span class="comment"># 判断总的variant的数目，用于得到正确的坐标</span></span><br><span class="line">            flag=<span class="number">1</span></span><br><span class="line">            <span class="comment"># 判断snv类型是不是存在，如果存在indel就会在上层，如果不存在indel就会出现在下层(存在第三种突变的情况)</span></span><br><span class="line">            flag_log=<span class="number">0</span></span><br><span class="line">            <span class="comment"># 因为原始的数据中，snv出现在indel之前，所以在indel上增加各种操作会比较麻烦</span></span><br><span class="line">            <span class="keyword">for</span> (i <span class="keyword">in</span> names(v))&#123;</span><br><span class="line">              <span class="keyword">if</span> (i ==<span class="string">"snv"</span> &amp; v[i]) &#123;</span><br><span class="line">                grid.rect(x, y - h*<span class="number">0.5</span> + flag/n*h, w*<span class="number">0.9</span>, <span class="number">1</span>/n*h, </span><br><span class="line">                                              gp = gpar(fill = <span class="literal">NA</span>, lwd = <span class="number">2</span>), just = <span class="string">"top"</span>)</span><br><span class="line">                <span class="comment"># 在snv上实现画斜线还是很简单的</span></span><br><span class="line">                grid.segments(x - w*<span class="number">0.9</span>*<span class="number">0.5</span>, </span><br><span class="line">                              y - h*<span class="number">0.5</span>,</span><br><span class="line">                              x + w*<span class="number">0.9</span>*<span class="number">0.5</span>, </span><br><span class="line">                              y - h*<span class="number">0.5</span> + <span class="number">1</span>/n*h, </span><br><span class="line">                              gp = gpar(lwd = <span class="number">2</span>))</span><br><span class="line">                flag=flag+<span class="number">1</span></span><br><span class="line">                flag_log=<span class="number">1</span></span><br><span class="line">              &#125;</span><br><span class="line">              <span class="comment"># 在indel上实现画斜线就很复杂</span></span><br><span class="line">              <span class="keyword">else</span> <span class="keyword">if</span>(flag_log)&#123;</span><br><span class="line">                <span class="keyword">if</span> (i ==<span class="string">"indel"</span> &amp; v[i] &amp; n==<span class="number">2</span>) &#123;</span><br><span class="line">                  grid.segments(x - w*<span class="number">0.9</span>*<span class="number">0.5</span>, </span><br><span class="line">                                y - h*<span class="number">0.5</span> + <span class="number">1</span>/n*h,</span><br><span class="line">                                x + w*<span class="number">0.9</span>*<span class="number">0.5</span>, </span><br><span class="line">                                y + h*<span class="number">0.5</span>, </span><br><span class="line">                                gp = gpar(lwd = <span class="number">2</span>))</span><br><span class="line">              &#125;</span><br><span class="line">                <span class="keyword">else</span> <span class="keyword">if</span> (i ==<span class="string">"indel"</span> &amp; v[i]) &#123;</span><br><span class="line">                  grid.segments(x - w*<span class="number">0.9</span>*<span class="number">0.5</span>, </span><br><span class="line">                                y - h*<span class="number">0.5</span> + <span class="number">1</span>/n*h,</span><br><span class="line">                                x + w*<span class="number">0.9</span>*<span class="number">0.5</span>, </span><br><span class="line">                                y + h*<span class="number">0.5</span> - <span class="number">1</span>/n*h, </span><br><span class="line">                                gp = gpar(lwd = <span class="number">2</span>))</span><br><span class="line">                &#125;</span><br><span class="line">              &#125;</span><br><span class="line">              <span class="keyword">else</span> <span class="keyword">if</span>(flag_log==<span class="number">0</span>)&#123;</span><br><span class="line">                <span class="keyword">if</span> (i ==<span class="string">"indel"</span> &amp; v[i] &amp; n==<span class="number">1</span>) &#123;</span><br><span class="line">                  grid.segments(x - w*<span class="number">0.9</span>*<span class="number">0.5</span>, </span><br><span class="line">                                y - h*<span class="number">0.5</span>,</span><br><span class="line">                                x + w*<span class="number">0.9</span>*<span class="number">0.5</span>, </span><br><span class="line">                                y + h*<span class="number">0.5</span>, </span><br><span class="line">                                gp = gpar(lwd = <span class="number">2</span>))</span><br><span class="line">                &#125;</span><br><span class="line">                <span class="keyword">else</span> <span class="keyword">if</span> (i ==<span class="string">"indel"</span> &amp; v[i] &amp; n&gt;<span class="number">1</span>) &#123;</span><br><span class="line">                  grid.segments(x - w*<span class="number">0.9</span>*<span class="number">0.5</span>, </span><br><span class="line">                                y - h*<span class="number">0.5</span>,</span><br><span class="line">                                x + w*<span class="number">0.9</span>*<span class="number">0.5</span>, </span><br><span class="line">                                y - h*<span class="number">0.5</span> + <span class="number">1</span>/n*h, </span><br><span class="line">                                gp = gpar(lwd = <span class="number">2</span>))</span><br><span class="line">                &#125;</span><br><span class="line">              &#125;</span><br><span class="line">            &#125;</span><br><span class="line">          &#125;, col = col)</span><br></pre></td></tr></table></figure><p></p><p><img src="https://showteeth.oss-cn-beijing.aliyuncs.com/blog_img/complexheatmap_oncoprint_intron_exon_2.png" alt="complexheatmap_oncoprint_intron_exon_2.png"></p><p>在indel上仅添加框线会更加简单一些：<br></p><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line">oncoPrint(mat_list,</span><br><span class="line">          alter_fun = <span class="keyword">function</span>(x, y, w, h, v) &#123;</span><br><span class="line">            n = sum(v)  <span class="comment"># how many alterations for current gene in current sample</span></span><br><span class="line">            h = h*<span class="number">0.9</span></span><br><span class="line">            <span class="comment"># use `names(which(v))` to correctly map between `v` and `col`</span></span><br><span class="line">            <span class="keyword">if</span>(n) grid.rect(x, y - h*<span class="number">0.5</span> + <span class="number">1</span>:n/n*h, w*<span class="number">0.9</span>, <span class="number">1</span>/n*h, </span><br><span class="line">                            gp = gpar(fill = col[names(which(v))], col = <span class="literal">NA</span>), just = <span class="string">"top"</span>)</span><br><span class="line">            flag=<span class="number">1</span></span><br><span class="line">            <span class="keyword">for</span> (i <span class="keyword">in</span> names(v))&#123;</span><br><span class="line">              <span class="keyword">if</span> (i ==<span class="string">"snv"</span> &amp; v[i]) &#123;</span><br><span class="line">                grid.rect(x, y - h*<span class="number">0.5</span> + flag/n*h, w*<span class="number">0.9</span>, <span class="number">1</span>/n*h, </span><br><span class="line">                                              gp = gpar(fill = <span class="literal">NA</span>, lwd = <span class="number">2</span>), just = <span class="string">"top"</span>)</span><br><span class="line">                grid.segments(x - w*<span class="number">0.9</span>*<span class="number">0.5</span>, </span><br><span class="line">                              y - h*<span class="number">0.5</span>,</span><br><span class="line">                              x + w*<span class="number">0.9</span>*<span class="number">0.5</span>, </span><br><span class="line">                              y - h*<span class="number">0.5</span> + <span class="number">1</span>/n*h, </span><br><span class="line">                              gp = gpar(lwd = <span class="number">2</span>))</span><br><span class="line">                flag=flag+<span class="number">1</span></span><br><span class="line">              &#125;</span><br><span class="line">              <span class="keyword">else</span> <span class="keyword">if</span> (i ==<span class="string">"indel"</span> &amp; v[i])&#123;</span><br><span class="line">                grid.rect(x, y - h*<span class="number">0.5</span> + flag/n*h, w*<span class="number">0.9</span>, <span class="number">1</span>/n*h, </span><br><span class="line">                          gp = gpar(fill = <span class="literal">NA</span>, lwd = <span class="number">2</span>), just = <span class="string">"top"</span>)</span><br><span class="line">              &#125;</span><br><span class="line">            &#125;</span><br><span class="line">          &#125;, col = col)</span><br></pre></td></tr></table></figure><p></p><p><img src="https://showteeth.oss-cn-beijing.aliyuncs.com/blog_img/complexheatmap_oncoprint_intron_exon_3.png" alt="complexheatmap_oncoprint_intron_exon_3.png"></p><hr><h3 id="其他相关参数"><a href="#其他相关参数" class="headerlink" title="其他相关参数"></a>其他相关参数</h3><ul><li><strong>控制列名(通常是样本名)</strong>：默认情况是<strong>不会显示</strong>的，可以通过参数<code>show_column_names = TRUE</code>来显示</li><li><strong>控制行名(通常是基因名)</strong>：默认情况是<strong>显示</strong>的，可以通过参数<code>show_row_names</code>来控制</li><li><strong>控制百分比</strong>：默认情况是显示的，可以通过参数<code>show_pct</code>来控制</li><li><strong>控制行名和百分比的显示方位</strong>：<code>pct_side</code>、<code>row_names_side</code></li><li><strong>控制百分比的小数点位数</strong>：<code>pct_digits</code></li><li><strong>控制barplot</strong>：<code>anno_oncoprint_barplot()</code>，具体示例请看帮助文档。</li><li><code>oncoPrint</code>也是热图，所以<strong>所有热图的参数这里也能使用</strong>，比如控制<code>legend</code>的参数<code>heatmap_legend_param</code></li><li><strong>删除没有突变的行或者列</strong>：默认情况下即使某行或者某列是没有突变也会显示出来，但是可以通过设置<code>remove_empty_columns</code>和<code>remove_empty_rows</code>为<code>TRUE</code>来进行删除。<strong>删除空行或者空列的操作最好在使用<code>oncoPrint</code>函数之前进行操作</strong>，防止在将<code>oncoPrint</code>结果和其他热图进行并列显示的时候出现问题。</li><li><strong>对行或者列重新排序</strong>：和前面热图的函数相同，可以使用<code>row_order</code>和<code>column_order</code>来控制进行控制。s</li></ul><h3 id="复杂oncoprint示例"><a href="#复杂oncoprint示例" class="headerlink" title="复杂oncoprint示例"></a>复杂oncoprint示例</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 输入数据准备</span></span><br><span class="line">mat = read.table(system.file(<span class="string">"extdata"</span>, package = <span class="string">"ComplexHeatmap"</span>, </span><br><span class="line">                             <span class="string">"tcga_lung_adenocarcinoma_provisional_ras_raf_mek_jnk_signalling.txt"</span>), </span><br><span class="line">                 header = <span class="literal">TRUE</span>, stringsAsFactors = <span class="literal">FALSE</span>, sep = <span class="string">"\t"</span>)</span><br><span class="line">mat[is.na(mat)] = <span class="string">""</span></span><br><span class="line">rownames(mat) = mat[, <span class="number">1</span>]</span><br><span class="line">mat = mat[, -<span class="number">1</span>]</span><br><span class="line">mat=  mat[, -ncol(mat)]</span><br><span class="line">mat = t(as.matrix(mat))</span><br><span class="line">mat[<span class="number">1</span>:<span class="number">3</span>, <span class="number">1</span>:<span class="number">3</span>]</span><br><span class="line">            TCGA-<span class="number">05</span>-<span class="number">4384</span>-<span class="number">01</span> TCGA-<span class="number">05</span>-<span class="number">4390</span>-<span class="number">01</span> TCGA-<span class="number">05</span>-<span class="number">4425</span>-<span class="number">01</span></span><br><span class="line">        KRAS <span class="string">"  "</span>            <span class="string">"MUT;"</span>          <span class="string">"  "</span>           </span><br><span class="line">        HRAS <span class="string">"  "</span>            <span class="string">"  "</span>            <span class="string">"  "</span>           </span><br><span class="line">        BRAF <span class="string">"  "</span>            <span class="string">"  "</span>            <span class="string">"  "</span>   </span><br><span class="line"><span class="comment"># 设置alter_fun</span></span><br><span class="line">col = c(<span class="string">"HOMDEL"</span> = <span class="string">"blue"</span>, <span class="string">"AMP"</span> = <span class="string">"red"</span>, <span class="string">"MUT"</span> = <span class="string">"#008000"</span>)</span><br><span class="line">alter_fun = list(</span><br><span class="line">  background = <span class="keyword">function</span>(x, y, w, h) &#123;</span><br><span class="line">    grid.rect(x, y, w-unit(<span class="number">0.5</span>, <span class="string">"mm"</span>), h-unit(<span class="number">0.5</span>, <span class="string">"mm"</span>), </span><br><span class="line">              gp = gpar(fill = <span class="string">"#CCCCCC"</span>, col = <span class="literal">NA</span>))</span><br><span class="line">  &#125;,</span><br><span class="line">  <span class="comment"># big blue</span></span><br><span class="line">  HOMDEL = <span class="keyword">function</span>(x, y, w, h) &#123;</span><br><span class="line">    grid.rect(x, y, w-unit(<span class="number">0.5</span>, <span class="string">"mm"</span>), h-unit(<span class="number">0.5</span>, <span class="string">"mm"</span>), </span><br><span class="line">              gp = gpar(fill = col[<span class="string">"HOMDEL"</span>], col = <span class="literal">NA</span>))</span><br><span class="line">  &#125;,</span><br><span class="line">  <span class="comment"># bug red</span></span><br><span class="line">  AMP = <span class="keyword">function</span>(x, y, w, h) &#123;</span><br><span class="line">    grid.rect(x, y, w-unit(<span class="number">0.5</span>, <span class="string">"mm"</span>), h-unit(<span class="number">0.5</span>, <span class="string">"mm"</span>), </span><br><span class="line">              gp = gpar(fill = col[<span class="string">"AMP"</span>], col = <span class="literal">NA</span>))</span><br><span class="line">  &#125;,</span><br><span class="line">  <span class="comment"># small green</span></span><br><span class="line">  MUT = <span class="keyword">function</span>(x, y, w, h) &#123;</span><br><span class="line">    grid.rect(x, y, w-unit(<span class="number">0.5</span>, <span class="string">"mm"</span>), h*<span class="number">0.33</span>, </span><br><span class="line">              gp = gpar(fill = col[<span class="string">"MUT"</span>], col = <span class="literal">NA</span>))</span><br><span class="line">  &#125;</span><br><span class="line">)</span><br><span class="line"></span><br><span class="line">column_title = <span class="string">"OncoPrint for TCGA Lung Adenocarcinoma, genes in Ras Raf MEK JNK signalling"</span></span><br><span class="line">heatmap_legend_param = list(title = <span class="string">"Alternations"</span>, at = c(<span class="string">"HOMDEL"</span>, <span class="string">"AMP"</span>, <span class="string">"MUT"</span>), </span><br><span class="line">                            labels = c(<span class="string">"Deep deletion"</span>, <span class="string">"Amplification"</span>, <span class="string">"Mutation"</span>))</span><br><span class="line"><span class="comment"># 绘图</span></span><br><span class="line">oncoPrint(mat,</span><br><span class="line">          alter_fun = alter_fun, col = col, </span><br><span class="line">          remove_empty_columns = <span class="literal">TRUE</span>, remove_empty_rows = <span class="literal">TRUE</span>,</span><br><span class="line">          <span class="comment"># 这里添加了三个注释，一个是bar图</span></span><br><span class="line">          <span class="comment"># 一个是1到172的热图，其实172是样本的数目</span></span><br><span class="line">          top_annotation = HeatmapAnnotation(cbar = anno_oncoprint_barplot(),</span><br><span class="line">                                             foo1 = <span class="number">1</span>:<span class="number">172</span>,</span><br><span class="line">                                             bar1 = anno_points(<span class="number">1</span>:<span class="number">172</span>)),</span><br><span class="line">          left_annotation = rowAnnotation(foo2 = <span class="number">1</span>:<span class="number">26</span>),</span><br><span class="line">          <span class="comment"># 26为基因的数目</span></span><br><span class="line">          right_annotation = rowAnnotation(bar2 = anno_barplot(<span class="number">1</span>:<span class="number">26</span>)),</span><br><span class="line">          column_title = column_title, heatmap_legend_param = heatmap_legend_param)</span><br></pre></td></tr></table></figure><p><img src="https://showteeth.oss-cn-beijing.aliyuncs.com/blog_img/complexheatmap_oncoprint_complex_1.png" alt="complexheatmap_oncoprint_complex_1.png"></p><hr><p><br></p><h2 id="参考链接"><a href="#参考链接" class="headerlink" title="参考链接"></a>参考链接</h2><ul><li><a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/index.html" target="_blank" rel="noopener">ComplexHeatmap-reference</a></li><li><a href="https://www.xiangyunhuang.com.cn/2019/08/grid/" target="_blank" rel="noopener">grid-栅格图形系统</a></li><li><a href="https://www.01hai.com/note/av133949" target="_blank" rel="noopener">R实战 第七篇：网格（grid）</a></li><li><a href="https://www.jianshu.com/nb/31400432" target="_blank" rel="noopener">ComplexHeatmap学习笔记系列</a></li><li><a href="https://www.jianshu.com/p/42adf881e587" target="_blank" rel="noopener">【r&lt;-包】ComplexHeatmap</a></li></ul><p><br></p><hr></div><div><div class="article_TheEnd">-----本文结束<i class="fa fa-paw"></i>感谢您的阅读-----</div></div><div><div class="my_post_copyright"><script src="//cdn.bootcss.com/clipboard.js/1.5.10/clipboard.min.js"></script><script src="https://cdn.bootcss.com/jquery/2.0.0/jquery.min.js"></script><script src="https://unpkg.com/sweetalert/dist/sweetalert.min.js"></script><p><span>本文标题:</span><a href="/posts/31202.html">R绘图系列-ComplexHeatmap使用记录</a></p><p><span>文章作者:</span><a href="/" title="访问 showteeth 的个人博客">showteeth</a></p><p><span>发布时间:</span>2019年12月19日 - 21:13</p><p><span>最后更新:</span>2020年05月18日 - 19:13</p><p><span>原始链接:</span><a href="/posts/31202.html" title="R绘图系列-ComplexHeatmap使用记录">http://showteeth.tech/posts/31202.html</a> <span class="copy-path" title="点击复制文章链接"><i class="fa fa-clipboard" data-clipboard-text="http://showteeth.tech/posts/31202.html" aria-label="复制成功！"></i></span></p><p><span>许可协议:</span><i class="fa fa-creative-commons"></i> <a rel="license" href="https://creativecommons.org/licenses/by-nc-nd/4.0/" target="_blank" title="Attribution-NonCommercial-NoDerivatives 4.0 International (CC BY-NC-ND 4.0)">署名-非商业性使用-禁止演绎 4.0 国际</a> 转载请保留原文链接及作者。</p></div><script>var clipboard=new Clipboard(".fa-clipboard");$(".fa-clipboard").click(function(){clipboard.on("success",function(){swal({title:"",text:"复制成功",icon:"success",showConfirmButton:!0})})})</script></div><footer class="post-footer"><div class="post-tags"><a href="/tags/绘图/" rel="tag"><i class="fa fa-tag"></i> 绘图</a></div><div class="post-nav"><div class="post-nav-next post-nav-item"><a href="/posts/55762.html" rel="next" title="R系列之在函数中保存图片"><i class="fa fa-chevron-left"></i> R系列之在函数中保存图片</a></div><span class="post-nav-divider"></span><div class="post-nav-prev post-nav-item"><a href="/posts/28801.html" rel="prev" title="shell系列之运算符优先级">shell系列之运算符优先级 <i class="fa fa-chevron-right"></i></a></div></div></footer></div></article><div class="post-spread"><script>window._bd_share_config={common:{bdText:"",bdMini:"1",bdMiniList:!1,bdPic:""},image:{viewList:["tsina","douban","sqq","qzone","weixin","twi","fbook"],viewText:"分享到：",viewSize:"16"},slide:{bdImg:"5",bdPos:"left",bdTop:"100"}}</script><script>with(document)(getElementsByTagName("head")[0]||body).appendChild(createElement("script")).src="//bdimg.share.baidu.com/static/api/js/share.js?cdnversion="+~(-new Date/36e5)</script></div></div></div><div class="comments" id="comments"></div></div><div class="sidebar-toggle"><div class="sidebar-toggle-line-wrap"><span class="sidebar-toggle-line sidebar-toggle-line-first"></span> <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span> <span class="sidebar-toggle-line sidebar-toggle-line-last"></span></div></div><aside id="sidebar" class="sidebar"><div class="sidebar-inner"><ul class="sidebar-nav motion-element"><li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">文章目录</li><li class="sidebar-nav-overview" data-target="site-overview-wrap">站点概览</li></ul><section class="site-overview-wrap sidebar-panel"><div class="site-overview"><div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person"><img class="site-author-image" itemprop="image" src="/images/timg.jfif" alt="showteeth"><p class="site-author-name" itemprop="name">showteeth</p><p class="site-description motion-element" itemprop="description"></p></div><nav class="site-state motion-element"><div class="site-state-item site-state-posts"><a href="/archives/"><span class="site-state-item-count">203</span> <span class="site-state-item-name">日志</span></a></div><div class="site-state-item site-state-categories"><a href="/categories/index.html"><span class="site-state-item-count">36</span> <span class="site-state-item-name">分类</span></a></div><div class="site-state-item site-state-tags"><a href="/tags/index.html"><span class="site-state-item-count">41</span> <span class="site-state-item-name">标签</span></a></div></nav><div class="feed-link motion-element"><a href="/atom.xml" rel="alternate"><i class="fa fa-rss"></i> RSS</a></div><div class="links-of-author motion-element"><span class="links-of-author-item"><a href="https://github.com/showteeth" target="_blank" title="GitHub"><i class="fa fa-fw fa-github"></i>GitHub</a></span></div><div class="links-of-blogroll motion-element links-of-blogroll-block"><div class="links-of-blogroll-title"><i class="fa fa-fw fa-link"></i> Links</div><ul class="links-of-blogroll-list"><li class="links-of-blogroll-item"><a href="http://example.com/" title="hexo" target="_blank">hexo</a></li></ul></div></div></section><section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active"><div class="post-toc"><div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#输入数据类型"><span class="nav-number">1.</span> <span class="nav-text">输入数据类型</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#设置颜色"><span class="nav-number">2.</span> <span class="nav-text">设置颜色</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#设置legend"><span class="nav-number">3.</span> <span class="nav-text">设置legend</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#设置row和col名称的顺序"><span class="nav-number">4.</span> <span class="nav-text">设置row和col名称的顺序</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#设置rowname的左右"><span class="nav-number">5.</span> <span class="nav-text">设置rowname的左右</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#设置cell的大小"><span class="nav-number">6.</span> <span class="nav-text">设置cell的大小</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#对某些行名和列名进行注释"><span class="nav-number">7.</span> <span class="nav-text">对某些行名和列名进行注释</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#在行或者列添加注释块"><span class="nav-number">8.</span> <span class="nav-text">在行或者列添加注释块</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#将不同的注释块分开"><span class="nav-number">9.</span> <span class="nav-text">将不同的注释块分开</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#完整脚本"><span class="nav-number">10.</span> <span class="nav-text">完整脚本</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#oncoPrint-适合展示突变"><span class="nav-number">11.</span> <span class="nav-text">oncoPrint-适合展示突变</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#输入数据"><span class="nav-number">11.1.</span> <span class="nav-text">输入数据</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#使用分隔符分割的多种突变名称"><span class="nav-number">11.1.1.</span> <span class="nav-text">使用分隔符分割的多种突变名称</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#使用0-1表示的多个variant列表"><span class="nav-number">11.1.2.</span> <span class="nav-text">使用0/1表示的多个variant列表</span></a></li></ol></li><li class="nav-item nav-level-3"><a class="nav-link" href="#得到突变类别"><span class="nav-number">11.2.</span> <span class="nav-text">得到突变类别</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#设置variant的颜色"><span class="nav-number">11.3.</span> <span class="nav-text">设置variant的颜色</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#设置variant的形状"><span class="nav-number">11.4.</span> <span class="nav-text">设置variant的形状</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#关于grid的位置参数"><span class="nav-number">11.4.1.</span> <span class="nav-text">关于grid的位置参数</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#按比例绘图"><span class="nav-number">11.4.2.</span> <span class="nav-text">按比例绘图</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#绘制三角形图案"><span class="nav-number">11.4.3.</span> <span class="nav-text">绘制三角形图案</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#用于支持更多variant类型"><span class="nav-number">11.4.4.</span> <span class="nav-text">用于支持更多variant类型</span></a></li></ol></li><li class="nav-item nav-level-3"><a class="nav-link" href="#设置grid背景颜色"><span class="nav-number">11.5.</span> <span class="nav-text">设置grid背景颜色</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#按重要性绘制图层"><span class="nav-number">11.6.</span> <span class="nav-text">按重要性绘制图层</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#其他相关参数"><span class="nav-number">11.7.</span> <span class="nav-text">其他相关参数</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#复杂oncoprint示例"><span class="nav-number">11.8.</span> <span class="nav-text">复杂oncoprint示例</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#参考链接"><span class="nav-number">12.</span> <span class="nav-text">参考链接</span></a></li></ol></div></div></section></div></aside></div></main><footer id="footer" class="footer"><div class="footer-inner"><script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script><div class="copyright">&copy; 2019 &mdash; <span itemprop="copyrightYear">2020</span> <span class="with-love"><i class="fa fa-user-secret"></i> </span><span class="author" itemprop="copyrightHolder">showteeth</span> <span class="post-meta-divider">|</span> <span class="post-meta-item-icon"><i class="fa fa-area-chart"></i> </span><span class="post-meta-item-text">博客全站共&#58;</span> <span title="Site words total count">380.2k 字</span></div><div><span id="timeDate" title="网站运行时间">载入天数...</span><span id="times" title="网站运行时间">载入时分秒...</span></div><script>var now=new Date;function createtime(){var n=new Date("02/15/2019 00:00:00");now.setTime(now.getTime()+250),days=(now-n)/1e3/60/60/24,dnum=Math.floor(days),hours=(now-n)/1e3/60/60-24*dnum,hnum=Math.floor(hours),1==String(hnum).length&&(hnum="0"+hnum),minutes=(now-n)/1e3/60-1440*dnum-60*hnum,mnum=Math.floor(minutes),1==String(mnum).length&&(mnum="0"+mnum),seconds=(now-n)/1e3-86400*dnum-3600*hnum-60*mnum,snum=Math.round(seconds),1==String(snum).length&&(snum="0"+snum),document.getElementById("timeDate").innerHTML="————网站已运行"+dnum+"&thinsp;天",document.getElementById("times").innerHTML=hnum+"&thinsp;时"+mnum+"&thinsp;分"+snum+"&thinsp;秒————"}setInterval("createtime()",250)</script><div class="busuanzi-count"><script async src="https://dn-lbstatics.qbox.me/busuanzi/2.3/busuanzi.pure.mini.js"></script><span class="site-uv"><i class="fa fa-user"></i> 本站访客数 <span class="busuanzi-value" id="busuanzi_value_site_uv"></span> 人次 </span><span class="site-pv"><i class="fa fa-eye"></i> 本站总访问量 <span class="busuanzi-value" id="busuanzi_value_site_pv"></span> 次</span></div></div></footer><div class="back-to-top"><i class="fa fa-arrow-up"></i> <span id="scrollpercent"><span>0</span>%</span></div></div><script type="text/javascript">"[object Function]"!==Object.prototype.toString.call(window.Promise)&&(window.Promise=null)</script><script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script><script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script><script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script><script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script><script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script><script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script><script type="text/javascript" src="/js/src/utils.js?v=5.1.4"></script><script type="text/javascript" src="/js/src/motion.js?v=5.1.4"></script><script type="text/javascript" src="/js/src/scrollspy.js?v=5.1.4"></script><script type="text/javascript" src="/js/src/post-details.js?v=5.1.4"></script><script type="text/javascript" src="/js/src/photoswipe.min.js?v=5.1.4"></script><script type="text/javascript" src="/js/src/photoswipe-ui-default.min.js?v=5.1.4"></script><script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.4"></script><script src="//cdn1.lncld.net/static/js/3.0.4/av-min.js"></script><script src="//unpkg.com/valine/dist/Valine.min.js"></script><script type="text/javascript">var GUEST = ['nick','mail','link'];
    var guest = 'nick,mail,link';
    guest = guest.split(',').filter(item=>{
      return GUEST.indexOf(item)>-1;
    });
    new Valine({
        el: '#comments' ,
        verify: false,
        notify: false,
        appId: 'RVmtcwupIO6WlzPf3zKfb2EW-gzGzoHsz',
        appKey: 'E8jip4oe60SRcVTMAbhG9cyY',
        placeholder: '快来评论吧~',
        avatar:'mm',
        guest_info:guest,
        pageSize:'10' || 10,
    });</script><script type="text/javascript">// Popup Window;
    var isfetched = false;
    var isXml = true;
    // Search DB path;
    var search_path = "search.xml";
    if (search_path.length === 0) {
      search_path = "search.xml";
    } else if (/json$/i.test(search_path)) {
      isXml = false;
    }
    var path = "/" + search_path;
    // monitor main search box;

    var onPopupClose = function (e) {
      $('.popup').hide();
      $('#local-search-input').val('');
      $('.search-result-list').remove();
      $('#no-result').remove();
      $(".local-search-pop-overlay").remove();
      $('body').css('overflow', '');
    }

    function proceedsearch() {
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay"></div>')
        .css('overflow', 'hidden');
      $('.search-popup-overlay').click(onPopupClose);
      $('.popup').toggle();
      var $localSearchInput = $('#local-search-input');
      $localSearchInput.attr("autocapitalize", "none");
      $localSearchInput.attr("autocorrect", "off");
      $localSearchInput.focus();
    }

    // search function;
    var searchFunc = function(path, search_id, content_id) {
      'use strict';

      // start loading animation
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay">' +
          '<div id="search-loading-icon">' +
          '<i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>' +
          '</div>' +
          '</div>')
        .css('overflow', 'hidden');
      $("#search-loading-icon").css('margin', '20% auto 0 auto').css('text-align', 'center');

      $.ajax({
        url: path,
        dataType: isXml ? "xml" : "json",
        async: true,
        success: function(res) {
          // get the contents from search data
          isfetched = true;
          $('.popup').detach().appendTo('.header-inner');
          var datas = isXml ? $("entry", res).map(function() {
            return {
              title: $("title", this).text(),
              content: $("content",this).text(),
              url: $("url" , this).text()
            };
          }).get() : res;
          var input = document.getElementById(search_id);
          var resultContent = document.getElementById(content_id);
          var inputEventFunction = function() {
            var searchText = input.value.trim().toLowerCase();
            var keywords = searchText.split(/[\s\-]+/);
            if (keywords.length > 1) {
              keywords.push(searchText);
            }
            var resultItems = [];
            if (searchText.length > 0) {
              // perform local searching
              datas.forEach(function(data) {
                var isMatch = false;
                var hitCount = 0;
                var searchTextCount = 0;
                var title = data.title.trim();
                var titleInLowerCase = title.toLowerCase();
                var content = data.content.trim().replace(/<[^>]+>/g,"");
                var contentInLowerCase = content.toLowerCase();
                var articleUrl = decodeURIComponent(data.url);
                var indexOfTitle = [];
                var indexOfContent = [];
                // only match articles with not empty titles
                if(title != '') {
                  keywords.forEach(function(keyword) {
                    function getIndexByWord(word, text, caseSensitive) {
                      var wordLen = word.length;
                      if (wordLen === 0) {
                        return [];
                      }
                      var startPosition = 0, position = [], index = [];
                      if (!caseSensitive) {
                        text = text.toLowerCase();
                        word = word.toLowerCase();
                      }
                      while ((position = text.indexOf(word, startPosition)) > -1) {
                        index.push({position: position, word: word});
                        startPosition = position + wordLen;
                      }
                      return index;
                    }

                    indexOfTitle = indexOfTitle.concat(getIndexByWord(keyword, titleInLowerCase, false));
                    indexOfContent = indexOfContent.concat(getIndexByWord(keyword, contentInLowerCase, false));
                  });
                  if (indexOfTitle.length > 0 || indexOfContent.length > 0) {
                    isMatch = true;
                    hitCount = indexOfTitle.length + indexOfContent.length;
                  }
                }

                // show search results

                if (isMatch) {
                  // sort index by position of keyword

                  [indexOfTitle, indexOfContent].forEach(function (index) {
                    index.sort(function (itemLeft, itemRight) {
                      if (itemRight.position !== itemLeft.position) {
                        return itemRight.position - itemLeft.position;
                      } else {
                        return itemLeft.word.length - itemRight.word.length;
                      }
                    });
                  });

                  // merge hits into slices

                  function mergeIntoSlice(text, start, end, index) {
                    var item = index[index.length - 1];
                    var position = item.position;
                    var word = item.word;
                    var hits = [];
                    var searchTextCountInSlice = 0;
                    while (position + word.length <= end && index.length != 0) {
                      if (word === searchText) {
                        searchTextCountInSlice++;
                      }
                      hits.push({position: position, length: word.length});
                      var wordEnd = position + word.length;

                      // move to next position of hit

                      index.pop();
                      while (index.length != 0) {
                        item = index[index.length - 1];
                        position = item.position;
                        word = item.word;
                        if (wordEnd > position) {
                          index.pop();
                        } else {
                          break;
                        }
                      }
                    }
                    searchTextCount += searchTextCountInSlice;
                    return {
                      hits: hits,
                      start: start,
                      end: end,
                      searchTextCount: searchTextCountInSlice
                    };
                  }

                  var slicesOfTitle = [];
                  if (indexOfTitle.length != 0) {
                    slicesOfTitle.push(mergeIntoSlice(title, 0, title.length, indexOfTitle));
                  }

                  var slicesOfContent = [];
                  while (indexOfContent.length != 0) {
                    var item = indexOfContent[indexOfContent.length - 1];
                    var position = item.position;
                    var word = item.word;
                    // cut out 100 characters
                    var start = position - 20;
                    var end = position + 80;
                    if(start < 0){
                      start = 0;
                    }
                    if (end < position + word.length) {
                      end = position + word.length;
                    }
                    if(end > content.length){
                      end = content.length;
                    }
                    slicesOfContent.push(mergeIntoSlice(content, start, end, indexOfContent));
                  }

                  // sort slices in content by search text's count and hits' count

                  slicesOfContent.sort(function (sliceLeft, sliceRight) {
                    if (sliceLeft.searchTextCount !== sliceRight.searchTextCount) {
                      return sliceRight.searchTextCount - sliceLeft.searchTextCount;
                    } else if (sliceLeft.hits.length !== sliceRight.hits.length) {
                      return sliceRight.hits.length - sliceLeft.hits.length;
                    } else {
                      return sliceLeft.start - sliceRight.start;
                    }
                  });

                  // select top N slices in content

                  var upperBound = parseInt('-1');
                  if (upperBound >= 0) {
                    slicesOfContent = slicesOfContent.slice(0, upperBound);
                  }

                  // highlight title and content

                  function highlightKeyword(text, slice) {
                    var result = '';
                    var prevEnd = slice.start;
                    slice.hits.forEach(function (hit) {
                      result += text.substring(prevEnd, hit.position);
                      var end = hit.position + hit.length;
                      result += '<b class="search-keyword">' + text.substring(hit.position, end) + '</b>';
                      prevEnd = end;
                    });
                    result += text.substring(prevEnd, slice.end);
                    return result;
                  }

                  var resultItem = '';

                  if (slicesOfTitle.length != 0) {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + highlightKeyword(title, slicesOfTitle[0]) + "</a>";
                  } else {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + title + "</a>";
                  }

                  slicesOfContent.forEach(function (slice) {
                    resultItem += "<a href='" + articleUrl + "'>" +
                      "<p class=\"search-result\">" + highlightKeyword(content, slice) +
                      "...</p>" + "</a>";
                  });

                  resultItem += "</li>";
                  resultItems.push({
                    item: resultItem,
                    searchTextCount: searchTextCount,
                    hitCount: hitCount,
                    id: resultItems.length
                  });
                }
              })
            };
            if (keywords.length === 1 && keywords[0] === "") {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-search fa-5x" /></div>'
            } else if (resultItems.length === 0) {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-frown-o fa-5x" /></div>'
            } else {
              resultItems.sort(function (resultLeft, resultRight) {
                if (resultLeft.searchTextCount !== resultRight.searchTextCount) {
                  return resultRight.searchTextCount - resultLeft.searchTextCount;
                } else if (resultLeft.hitCount !== resultRight.hitCount) {
                  return resultRight.hitCount - resultLeft.hitCount;
                } else {
                  return resultRight.id - resultLeft.id;
                }
              });
              var searchResultList = '<ul class=\"search-result-list\">';
              resultItems.forEach(function (result) {
                searchResultList += result.item;
              })
              searchResultList += "</ul>";
              resultContent.innerHTML = searchResultList;
            }
          }

          if ('auto' === 'auto') {
            input.addEventListener('input', inputEventFunction);
          } else {
            $('.search-icon').click(inputEventFunction);
            input.addEventListener('keypress', function (event) {
              if (event.keyCode === 13) {
                inputEventFunction();
              }
            });
          }

          // remove loading animation
          $(".local-search-pop-overlay").remove();
          $('body').css('overflow', '');

          proceedsearch();
        }
      });
    }

    // handle and trigger popup window;
    $('.popup-trigger').click(function(e) {
      e.stopPropagation();
      if (isfetched === false) {
        searchFunc(path, 'local-search-input', 'local-search-result');
      } else {
        proceedsearch();
      };
    });

    $('.popup-btn-close').click(onPopupClose);
    $('.popup').click(function(e){
      e.stopPropagation();
    });
    $(document).on('keyup', function (event) {
      var shouldDismissSearchPopup = event.which === 27 &&
        $('.search-popup').is(':visible');
      if (shouldDismissSearchPopup) {
        onPopupClose();
      }
    });</script><script type="text/javascript" src="/js/src/clipboard.min.js"></script><script type="text/javascript" src="/js/src/clipboard-use.js"></script><script src="/js/src/photoswipe.min.js?v=5.1.4"></script><script src="/js/src/photoswipe-ui-default.min.js?v=5.1.4"></script></body></html>